// @jeepq/core/dist/collection/components/jeep-slides/jeep-slides.js
import { h, Host } from "@stencil/core";
import { getBoundingClientRect } from '../../utils/common';
export class JeepSlides {
    constructor() {
        this.fullScreen = false;
        this._slidesParams = {};
        this._ticking = false;
        this._oriMargin = {};
        this._touchStart = false;
        this._touchMove = false;
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    parseOptionsProp(newValue) {
        if (newValue)
            this.innerOptions = JSON.parse(newValue);
    }
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    handleWindowResize() {
        this._windowResize();
    }
    async handleJeepNavigationPrev() {
        await this.goToPreviousSlide();
    }
    async handleJeepNavigationNext() {
        await this.goToNextSlide();
    }
    async handleJeepPaginationIndex(event) {
        if (event.detail && event.detail.index) {
            await this.goToSlideIndex(Number(event.detail.index));
            if (this.innerOptions.navigation)
                this._setNavigationButtons();
            await this._pagEl.setJeepPaginationActiveIndex({ activeIndex: this._slidesParams.curSlide.toString() });
        }
    }
    async handleJeepPlayControlsCurrentIndex(event) {
        if (event.detail && event.detail.index) {
            await this.goToSlideIndex(Number(event.detail.index));
        }
    }
    async handleJeepFullscreenRequest() {
        this._setProperties(false);
        await this._fullEl.fullscreenRequest(document.documentElement);
        this.fullScreen = true;
        this._windowResize();
    }
    async handleJeepFullscreenExit() {
        await this._fullEl.fullscreenExit();
        this._setProperties(true);
        this.fullScreen = false;
        this._windowResize();
    }
    async handleJeepFullscreenChange() {
        this._setProperties(true);
        this.fullScreen = false;
        this._windowResize();
    }
    async handleKeydown(event) {
        event.preventDefault();
        const key = event.key ? event.key : null;
        const keycode = event.keyCode || event.which ? event.keyCode || event.which : null;
        const pressKey = key !== null ? key : keycode.toString();
        switch (pressKey) {
            case 'ArrowLeft' || '37': // LEFT
                if (this._direction === 'horizontal') {
                    await this.goToPreviousSlide();
                }
                break;
            case 'ArrowUp' || '38': // UP
                if (this._direction === 'vertical') {
                    await this.goToPreviousSlide();
                }
                break;
            case 'ArrowRight' || '39': // RIGHT
                if (this._direction === 'horizontal') {
                    await this.goToNextSlide();
                }
                break;
            case 'ArrowDown' || '40': // DOWN
                if (this._direction === 'vertical') {
                    await this.goToNextSlide();
                }
                break;
            default:
        }
    }
    handleTouchstart() {
        this._touchStart = true;
    }
    handleTouchmove() {
        this._touchMove = true;
    }
    async handleTouchend() {
        if (!this._touchStart || !this._touchMove) {
            this._initial = await this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500);
        }
        this._touchStart = false;
        this._touchMove = false;
    }
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    init() {
        return Promise.resolve(this._init());
    }
    /**
     * Set the slides.
     */
    setSlides() {
        return Promise.resolve(this._setSlides());
    }
    /**
     * Get Active Slide Index.
     */
    async getActiveSlideIndex() {
        const curSlide = await this._getCurrentSlideIndex();
        return Promise.resolve(curSlide);
    }
    /**
     *  Go Previous Slide
     */
    async goToPreviousSlide() {
        return await this._goPrev();
    }
    /**
     *  Go Next Slide
     */
    async goToNextSlide() {
        return await this._goNext();
    }
    /**
     *  Go To Slide with given index
     */
    async goToSlideIndex(index) {
        return await this._scrollToPosition(index);
    }
    async componentWillLoad() {
        await this.init();
    }
    async componentDidLoad() {
        await this.setSlides();
    }
    async _init() {
        this._element = this.el.shadowRoot;
        this._slidesNumber = 0;
        const defOpt = { direction: "horizontal" };
        this.parseOptionsProp(this.options ? this.options : JSON.stringify(defOpt));
        this._direction = this.innerOptions.direction ? this.innerOptions.direction : "horizontal";
        this._navIcon = this.innerOptions.navigation && this.innerOptions.navigation.icon ? this.innerOptions.navigation.icon : "arrow-circle";
        this._oriMargin.top = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-top');
        this._oriMargin.left = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-left');
        this._oriMargin.bottom = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-bottom');
        this._oriMargin.right = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-right');
    }
    async _setSlides() {
        this._containerEl = this._element.querySelector('.slides-container');
        this._wrapperEl = this._containerEl.querySelector('.slides-wrapper');
        this._slidesEl = this._wrapperEl.querySelector('.slides');
        this._slidesEl.classList.add(`slides-${this._direction}`);
        this._slotEl = this._slidesEl.querySelector('slot');
        if (this._slotEl !== null) {
            this._arrSlidesEl = await this._getElementsInSlot();
            if (this._arrSlidesEl !== null && this._arrSlidesEl.length > 0) {
                //<-- required to be added as the scroll horizontal does not work with slot element
                // loop through the slot element , add the elements to the render
                // store elements which are in slot
                this._slidesNumber = this._arrSlidesEl.length;
                for (let i = 0; i < this._slidesNumber; i++) {
                    this._slidesEl.append(this._arrSlidesEl[i]);
                }
                // delete the slot element
                const div = Array.from(this.el.querySelectorAll('div'));
                if (div !== null) {
                    for (let i = 0; i < div.length; i++) {
                        if (div[i].slot === "slides")
                            this.el.removeChild(div[i]);
                    }
                    this._slidesEl.removeChild(this._slotEl);
                }
                //-->
                if (this.innerOptions.fullscreen)
                    this._fullEl = this._containerEl.querySelector('#jeep-slides-fullscreen');
                if (!this.innerOptions.autoplay && this.innerOptions.navigation)
                    this._navEl = this._containerEl.querySelector('#jeep-slides-navigation');
                if (!this.innerOptions.autoplay && this.innerOptions.pagination) {
                    const pagId = `#jeep-slides-pagination-${this.innerOptions.direction}`;
                    this._pagEl = this._containerEl.querySelector(pagId);
                    await this._pagEl.setJeepPaginationSlidesNumber({ slides: this._arrSlidesEl.length });
                }
                if (this.innerOptions.autoplay)
                    this._playEl = this._containerEl.querySelector('#jeep-slides-autoplay');
                // add scroll event
                this._slidesEl.addEventListener('scroll', () => {
                    const prevIndex = this._slidesParams.curSlide;
                    if (!this._ticking) {
                        setTimeout(async () => {
                            this._slidesParams.curSlide = await this.getActiveSlideIndex();
                            this._slidesParams.position = this._slidesParams.axis === 'x' ? Number(this._slidesEl.scrollLeft) : Number(this._slidesEl.scrollTop);
                            if (this.innerOptions.navigation && this._slidesParams.curSlide !== prevIndex)
                                this._setNavigationButtons();
                            if (this.innerOptions.pagination && this._slidesParams.curSlide !== prevIndex)
                                this._syncPagination();
                            if (this.innerOptions.autoplay && this._slidesParams.curSlide !== prevIndex)
                                this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
                            this._ticking = false;
                        }, 200);
                        this._ticking = true;
                    }
                }, false);
                // add click event
                this._slidesEl.addEventListener('click', async () => {
                    this._initial = await this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500);
                }, false);
                // fill the slidesParams
                this._slidesParams.axis = this._direction === 'vertical' ? 'y' : 'x';
                this._slidesParams.curSlide = 0;
                this._slidesParams.nbSlides = this._arrSlidesEl.length;
                await this._initScrollParams();
                if (!this.innerOptions.autoplay && (this.innerOptions.navigation && this._slidesParams.nbSlides > 1))
                    await this._navEl.setJeepNavigationNextDisabled({ disabled: false });
                this._initial = await this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500);
                this.nSlides = this._arrSlidesEl.length;
            }
        }
    }
    _setProperties(back) {
        if (!back) {
            this.el.style.setProperty('--slides-margin-top', '0px');
            this.el.style.setProperty('--slides-margin-bottom', '0px');
            this.el.style.setProperty('--slides-margin-left', '0px');
            this.el.style.setProperty('--slides-margin-right', '0px');
        }
        else {
            this.el.style.setProperty('--slides-margin-top', this._oriMargin.top);
            this.el.style.setProperty('--slides-margin-bottom', this._oriMargin.bottom);
            this.el.style.setProperty('--slides-margin-left', this._oriMargin.left);
            this.el.style.setProperty('--slides-margin-right', this._oriMargin.right);
        }
    }
    async _windowResize() {
        this.el.style.setProperty('--slides-width', `${window.innerWidth}px`);
        this.el.style.setProperty('--slides-height', `${window.innerHeight}px`);
        await this._initScrollParams();
        await this.goToSlideIndex(this._slidesParams.curSlide);
    }
    async _getElementsInSlot() {
        let elems = null;
        const nodes = this._slotEl.assignedNodes();
        const cmpSlides = nodes.filter((item) => {
            return item.slot === 'slides' ? item : null;
        });
        if (cmpSlides !== null && cmpSlides.length > 0) {
            elems = Array.from(cmpSlides[0].querySelectorAll('jeep-slide'));
        }
        return elems;
    }
    async _initScrollParams() {
        const rect = await getBoundingClientRect(this._arrSlidesEl[0], 100);
        this._slidesParams.slideLength = this._slidesParams.axis === 'x' ? rect.width : rect.height;
        this._slidesParams.scrollLength = this._slidesParams.axis === 'x' ? this._slidesEl.scrollWidth :
            this._slidesEl.scrollHeight;
        this._slidesParams.position = this._slidesParams.curSlide * this._slidesParams.slideLength;
    }
    async _doHide(duration) {
        clearTimeout(this._initial);
        await this._doVisible();
        let initial = setTimeout(async () => {
            this.onSlidesHeaderVisibility.emit({ visibility: 'hidden' });
            if (this._navEl && this.innerOptions.navigation.hidden)
                await this._navEl.setJeepNavigationVisibility({ visibility: 'hidden' });
            if (this._pagEl && this.innerOptions.pagination.hidden)
                await this._pagEl.setJeepPaginationVisibility({ visibility: 'hidden' });
            if (this._playEl && this.innerOptions.autoplay.hidden)
                await this._playEl.setJeepPlayControlsVisibility({ visibility: 'hidden' });
            if (this._fullEl && this.innerOptions.fullscreen.hidden)
                await this._fullEl.setJeepFullscreenVisibility({ visibility: 'hidden' });
        }, duration);
        return initial;
    }
    async _doVisible() {
        if (this._slidesParams.curSlide === 0 || this._slidesParams.curSlide === this._slidesNumber - 1)
            this.onSlidesHeaderVisibility.emit({ visibility: 'visible' });
        if (this._navEl && this.innerOptions.navigation.hidden)
            await this._navEl.setJeepNavigationVisibility({ visibility: 'visible' });
        if (this._pagEl && this.innerOptions.pagination.hidden)
            await this._pagEl.setJeepPaginationVisibility({ visibility: 'visible' });
        if (this._playEl && this.innerOptions.autoplay.hidden)
            await this._playEl.setJeepPlayControlsVisibility({ visibility: 'visible' });
        if (this._fullEl && this.innerOptions.fullscreen.hidden)
            await this._fullEl.setJeepFullscreenVisibility({ visibility: 'visible' });
        return;
    }
    async _getCurrentSlideIndex() {
        let index = 0;
        this._slidesParams.position = this._slidesParams.axis === 'x' ? Number(this._slidesEl.scrollLeft) : Number(this._slidesEl.scrollTop);
        if (Math.abs(this._slidesParams.scrollLength / this._slidesParams.nbSlides - this._slidesParams.slideLength) < 2) {
            // slide constant width
            index = Math.round(this._slidesParams.position / this._slidesParams.slideLength);
        }
        return index;
    }
    async _goPrev() {
        this._slidesParams.curSlide -= 1;
        if (this._slidesParams.curSlide < 0)
            this._slidesParams.curSlide = 0;
        if (this.innerOptions.autoplay) {
            this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
        }
        else {
            this._setNavigationButtons();
            if (this.innerOptions.pagination)
                this._syncPagination();
        }
        await this.goToSlideIndex(this._slidesParams.curSlide);
    }
    async _goNext() {
        this._slidesParams.curSlide += 1;
        if (this._slidesParams.curSlide >= this._slidesParams.nbSlides)
            this._slidesParams.curSlide = this._slidesParams.nbSlides - 1;
        if (this.innerOptions.autoplay) {
            this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
        }
        else {
            this._setNavigationButtons();
            if (this.innerOptions.pagination)
                this._syncPagination();
        }
        await this.goToSlideIndex(this._slidesParams.curSlide);
    }
    async _setNavigationButtons() {
        if (this._navEl) {
            if (this._slidesParams.curSlide === 0)
                await this._navEl.setJeepNavigationPrevDisabled({ disabled: true });
            if (this._slidesParams.curSlide >= 1)
                await this._navEl.setJeepNavigationPrevDisabled({ disabled: false });
            if (this._slidesParams.curSlide === this._slidesParams.nbSlides - 1)
                await this._navEl.setJeepNavigationNextDisabled({ disabled: true });
            if (this._slidesParams.curSlide <= this._slidesParams.nbSlides - 2)
                await this._navEl.setJeepNavigationNextDisabled({ disabled: false });
        }
    }
    _syncPagination() {
        if (this._pagEl)
            this._pagEl.setJeepPaginationActiveIndex({ activeIndex: this._slidesParams.curSlide.toString() });
    }
    async _scrollToPosition(index) {
        if (index !== this._slidesParams.curSlide)
            this._slidesParams.curSlide = index;
        this._slidesParams.position = this._slidesParams.curSlide * this._slidesParams.slideLength;
        if (this._slidesParams.axis === 'x') {
            if (this._slidesEl)
                this._slidesEl.scrollTo(this._slidesParams.position, 0);
        }
        else {
            if (this._slidesEl)
                this._slidesEl.scrollTo(0, this._slidesParams.position);
        }
    }
    render() {
        return (h(Host, null,
            h("div", { class: "slides-container" },
                h("div", { class: "slides-wrapper" },
                    h("div", { class: "slides" },
                        h("slot", { name: "slides" }))),
                !this.innerOptions.autoplay && this.innerOptions.navigation
                    ? h("jeep-navigation", { id: "jeep-slides-navigation", name: this._navIcon })
                    : null,
                !this.innerOptions.autoplay && this.innerOptions.pagination
                    ? h("jeep-pagination", { id: `jeep-slides-pagination-${this._direction}`, ndisplay: this.innerOptions.pagination.bulletsDisplay ? this.innerOptions.pagination.bulletsDisplay : 5, clickable: this.innerOptions.pagination.clickable ? this.innerOptions.pagination.clickable : false, direction: this._direction })
                    : null,
                this.innerOptions.autoplay
                    ? h("jeep-play-controls", { id: "jeep-slides-autoplay", duration: this.innerOptions.autoplay.duration, nslides: this.nSlides })
                    : null,
                this.innerOptions.fullscreen
                    ? h("jeep-fullscreen", { id: "jeep-slides-fullscreen" })
                    : null)));
    }
    static get is() { return "jeep-slides"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["jeep-slides.css"]
    }; }
    static get styleUrls() { return {
        "$": ["jeep-slides.css"]
    }; }
    static get properties() { return {
        "options": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The slides options"
            },
            "attribute": "options",
            "reflect": false
        }
    }; }
    static get states() { return {
        "innerOptions": {},
        "nSlides": {},
        "fullScreen": {}
    }; }
    static get events() { return [{
            "method": "onSlidesHeaderVisibility",
            "name": "jeepSlidesHeaderVisibility",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Emitted the Header visibility change"
            },
            "complexType": {
                "original": "HeaderVisibility",
                "resolved": "HeaderVisibility",
                "references": {
                    "HeaderVisibility": {
                        "location": "import",
                        "path": "../../global/interfaces/jeep-slides"
                    }
                }
            }
        }]; }
    static get methods() { return {
        "init": {
            "complexType": {
                "signature": "() => Promise<void>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "Init data from properties.",
                "tags": []
            }
        },
        "setSlides": {
            "complexType": {
                "signature": "() => Promise<void>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "Set the slides.",
                "tags": []
            }
        },
        "getActiveSlideIndex": {
            "complexType": {
                "signature": "() => Promise<number>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<number>"
            },
            "docs": {
                "text": "Get Active Slide Index.",
                "tags": []
            }
        }
    }; }
    static get elementRef() { return "el"; }
    static get watchers() { return [{
            "propName": "options",
            "methodName": "parseOptionsProp"
        }]; }
    static get listeners() { return [{
            "name": "resize",
            "method": "handleWindowResize",
            "target": "window",
            "capture": false,
            "passive": true
        }, {
            "name": "jeepNavigationPrev",
            "method": "handleJeepNavigationPrev",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "jeepNavigationNext",
            "method": "handleJeepNavigationNext",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "jeepPaginationIndex",
            "method": "handleJeepPaginationIndex",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "jeepPlayControlsCurrentIndex",
            "method": "handleJeepPlayControlsCurrentIndex",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "jeepFullscreenRequest",
            "method": "handleJeepFullscreenRequest",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "jeepFullscreenExit",
            "method": "handleJeepFullscreenExit",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "jeepFullscreenChange",
            "method": "handleJeepFullscreenChange",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "keydown",
            "method": "handleKeydown",
            "target": "document",
            "capture": false,
            "passive": false
        }, {
            "name": "touchstart",
            "method": "handleTouchstart",
            "target": undefined,
            "capture": false,
            "passive": true
        }, {
            "name": "touchmove",
            "method": "handleTouchmove",
            "target": undefined,
            "capture": false,
            "passive": true
        }, {
            "name": "touchend",
            "method": "handleTouchend",
            "target": undefined,
            "capture": false,
            "passive": true
        }]; }
}
