// @jeepq/core/dist/cjs/jeep-slides.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-ce02df39.js');
const common = require('./common-5e122d9c.js');

const JeepSlides = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        this.fullScreen = false;
        this._slidesParams = {};
        this._ticking = false;
        this._oriMargin = {};
        this._touchStart = false;
        this._touchMove = false;
        this.onSlidesHeaderVisibility = core.createEvent(this, "jeepSlidesHeaderVisibility", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    parseOptionsProp(newValue) {
        if (newValue)
            this.innerOptions = JSON.parse(newValue);
    }
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    handleWindowResize() {
        this._windowResize();
    }
    async handleJeepNavigationPrev() {
        await this.goToPreviousSlide();
    }
    async handleJeepNavigationNext() {
        await this.goToNextSlide();
    }
    async handleJeepPaginationIndex(event) {
        if (event.detail && event.detail.index) {
            await this.goToSlideIndex(Number(event.detail.index));
            if (this.innerOptions.navigation)
                this._setNavigationButtons();
            await this._pagEl.setJeepPaginationActiveIndex({ activeIndex: this._slidesParams.curSlide.toString() });
        }
    }
    async handleJeepPlayControlsCurrentIndex(event) {
        if (event.detail && event.detail.index) {
            await this.goToSlideIndex(Number(event.detail.index));
        }
    }
    async handleJeepFullscreenRequest() {
        this._setProperties(false);
        await this._fullEl.fullscreenRequest(document.documentElement);
        this.fullScreen = true;
        this._windowResize();
    }
    async handleJeepFullscreenExit() {
        await this._fullEl.fullscreenExit();
        this._setProperties(true);
        this.fullScreen = false;
        this._windowResize();
    }
    async handleJeepFullscreenChange() {
        this._setProperties(true);
        this.fullScreen = false;
        this._windowResize();
    }
    async handleKeydown(event) {
        event.preventDefault();
        const key = event.key ? event.key : null;
        const keycode = event.keyCode || event.which ? event.keyCode || event.which : null;
        const pressKey = key !== null ? key : keycode.toString();
        switch (pressKey) {
            case 'ArrowLeft' : // LEFT
                if (this._direction === 'horizontal') {
                    await this.goToPreviousSlide();
                }
                break;
            case 'ArrowUp' : // UP
                if (this._direction === 'vertical') {
                    await this.goToPreviousSlide();
                }
                break;
            case 'ArrowRight' : // RIGHT
                if (this._direction === 'horizontal') {
                    await this.goToNextSlide();
                }
                break;
            case 'ArrowDown' : // DOWN
                if (this._direction === 'vertical') {
                    await this.goToNextSlide();
                }
                break;
        }
    }
    handleTouchstart() {
        this._touchStart = true;
    }
    handleTouchmove() {
        this._touchMove = true;
    }
    async handleTouchend() {
        if (!this._touchStart || !this._touchMove) {
            this._initial = await this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500);
        }
        this._touchStart = false;
        this._touchMove = false;
    }
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    init() {
        return Promise.resolve(this._init());
    }
    /**
     * Set the slides.
     */
    setSlides() {
        return Promise.resolve(this._setSlides());
    }
    /**
     * Get Active Slide Index.
     */
    async getActiveSlideIndex() {
        const curSlide = await this._getCurrentSlideIndex();
        return Promise.resolve(curSlide);
    }
    /**
     *  Go Previous Slide
     */
    async goToPreviousSlide() {
        return await this._goPrev();
    }
    /**
     *  Go Next Slide
     */
    async goToNextSlide() {
        return await this._goNext();
    }
    /**
     *  Go To Slide with given index
     */
    async goToSlideIndex(index) {
        return await this._scrollToPosition(index);
    }
    async componentWillLoad() {
        await this.init();
    }
    async componentDidLoad() {
        await this.setSlides();
    }
    async _init() {
        this._element = this.el.shadowRoot;
        this._slidesNumber = 0;
        const defOpt = { direction: "horizontal" };
        this.parseOptionsProp(this.options ? this.options : JSON.stringify(defOpt));
        this._direction = this.innerOptions.direction ? this.innerOptions.direction : "horizontal";
        this._navIcon = this.innerOptions.navigation && this.innerOptions.navigation.icon ? this.innerOptions.navigation.icon : "arrow-circle";
        this._oriMargin.top = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-top');
        this._oriMargin.left = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-left');
        this._oriMargin.bottom = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-bottom');
        this._oriMargin.right = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-right');
    }
    async _setSlides() {
        this._containerEl = this._element.querySelector('.slides-container');
        this._wrapperEl = this._containerEl.querySelector('.slides-wrapper');
        this._slidesEl = this._wrapperEl.querySelector('.slides');
        this._slidesEl.classList.add(`slides-${this._direction}`);
        this._slotEl = this._slidesEl.querySelector('slot');
        if (this._slotEl !== null) {
            this._arrSlidesEl = await this._getElementsInSlot();
            if (this._arrSlidesEl !== null && this._arrSlidesEl.length > 0) {
                //<-- required to be added as the scroll horizontal does not work with slot element
                // loop through the slot element , add the elements to the render
                // store elements which are in slot
                this._slidesNumber = this._arrSlidesEl.length;
                for (let i = 0; i < this._slidesNumber; i++) {
                    this._slidesEl.append(this._arrSlidesEl[i]);
                }
                // delete the slot element
                const div = Array.from(this.el.querySelectorAll('div'));
                if (div !== null) {
                    for (let i = 0; i < div.length; i++) {
                        if (div[i].slot === "slides")
                            this.el.removeChild(div[i]);
                    }
                    this._slidesEl.removeChild(this._slotEl);
                }
                //-->
                if (this.innerOptions.fullscreen)
                    this._fullEl = this._containerEl.querySelector('#jeep-slides-fullscreen');
                if (!this.innerOptions.autoplay && this.innerOptions.navigation)
                    this._navEl = this._containerEl.querySelector('#jeep-slides-navigation');
                if (!this.innerOptions.autoplay && this.innerOptions.pagination) {
                    const pagId = `#jeep-slides-pagination-${this.innerOptions.direction}`;
                    this._pagEl = this._containerEl.querySelector(pagId);
                    await this._pagEl.setJeepPaginationSlidesNumber({ slides: this._arrSlidesEl.length });
                }
                if (this.innerOptions.autoplay)
                    this._playEl = this._containerEl.querySelector('#jeep-slides-autoplay');
                // add scroll event
                this._slidesEl.addEventListener('scroll', () => {
                    const prevIndex = this._slidesParams.curSlide;
                    if (!this._ticking) {
                        setTimeout(async () => {
                            this._slidesParams.curSlide = await this.getActiveSlideIndex();
                            this._slidesParams.position = this._slidesParams.axis === 'x' ? Number(this._slidesEl.scrollLeft) : Number(this._slidesEl.scrollTop);
                            if (this.innerOptions.navigation && this._slidesParams.curSlide !== prevIndex)
                                this._setNavigationButtons();
                            if (this.innerOptions.pagination && this._slidesParams.curSlide !== prevIndex)
                                this._syncPagination();
                            if (this.innerOptions.autoplay && this._slidesParams.curSlide !== prevIndex)
                                this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
                            this._ticking = false;
                        }, 200);
                        this._ticking = true;
                    }
                }, false);
                // add click event
                this._slidesEl.addEventListener('click', async () => {
                    this._initial = await this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500);
                }, false);
                // fill the slidesParams
                this._slidesParams.axis = this._direction === 'vertical' ? 'y' : 'x';
                this._slidesParams.curSlide = 0;
                this._slidesParams.nbSlides = this._arrSlidesEl.length;
                await this._initScrollParams();
                if (!this.innerOptions.autoplay && (this.innerOptions.navigation && this._slidesParams.nbSlides > 1))
                    await this._navEl.setJeepNavigationNextDisabled({ disabled: false });
                this._initial = await this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500);
                this.nSlides = this._arrSlidesEl.length;
            }
        }
    }
    _setProperties(back) {
        if (!back) {
            this.el.style.setProperty('--slides-margin-top', '0px');
            this.el.style.setProperty('--slides-margin-bottom', '0px');
            this.el.style.setProperty('--slides-margin-left', '0px');
            this.el.style.setProperty('--slides-margin-right', '0px');
        }
        else {
            this.el.style.setProperty('--slides-margin-top', this._oriMargin.top);
            this.el.style.setProperty('--slides-margin-bottom', this._oriMargin.bottom);
            this.el.style.setProperty('--slides-margin-left', this._oriMargin.left);
            this.el.style.setProperty('--slides-margin-right', this._oriMargin.right);
        }
    }
    async _windowResize() {
        this.el.style.setProperty('--slides-width', `${window.innerWidth}px`);
        this.el.style.setProperty('--slides-height', `${window.innerHeight}px`);
        await this._initScrollParams();
        await this.goToSlideIndex(this._slidesParams.curSlide);
    }
    async _getElementsInSlot() {
        let elems = null;
        const nodes = this._slotEl.assignedNodes();
        const cmpSlides = nodes.filter((item) => {
            return item.slot === 'slides' ? item : null;
        });
        if (cmpSlides !== null && cmpSlides.length > 0) {
            elems = Array.from(cmpSlides[0].querySelectorAll('jeep-slide'));
        }
        return elems;
    }
    async _initScrollParams() {
        const rect = await common.getBoundingClientRect(this._arrSlidesEl[0], 100);
        this._slidesParams.slideLength = this._slidesParams.axis === 'x' ? rect.width : rect.height;
        this._slidesParams.scrollLength = this._slidesParams.axis === 'x' ? this._slidesEl.scrollWidth :
            this._slidesEl.scrollHeight;
        this._slidesParams.position = this._slidesParams.curSlide * this._slidesParams.slideLength;
    }
    async _doHide(duration) {
        clearTimeout(this._initial);
        await this._doVisible();
        let initial = setTimeout(async () => {
            this.onSlidesHeaderVisibility.emit({ visibility: 'hidden' });
            if (this._navEl && this.innerOptions.navigation.hidden)
                await this._navEl.setJeepNavigationVisibility({ visibility: 'hidden' });
            if (this._pagEl && this.innerOptions.pagination.hidden)
                await this._pagEl.setJeepPaginationVisibility({ visibility: 'hidden' });
            if (this._playEl && this.innerOptions.autoplay.hidden)
                await this._playEl.setJeepPlayControlsVisibility({ visibility: 'hidden' });
            if (this._fullEl && this.innerOptions.fullscreen.hidden)
                await this._fullEl.setJeepFullscreenVisibility({ visibility: 'hidden' });
        }, duration);
        return initial;
    }
    async _doVisible() {
        if (this._slidesParams.curSlide === 0 || this._slidesParams.curSlide === this._slidesNumber - 1)
            this.onSlidesHeaderVisibility.emit({ visibility: 'visible' });
        if (this._navEl && this.innerOptions.navigation.hidden)
            await this._navEl.setJeepNavigationVisibility({ visibility: 'visible' });
        if (this._pagEl && this.innerOptions.pagination.hidden)
            await this._pagEl.setJeepPaginationVisibility({ visibility: 'visible' });
        if (this._playEl && this.innerOptions.autoplay.hidden)
            await this._playEl.setJeepPlayControlsVisibility({ visibility: 'visible' });
        if (this._fullEl && this.innerOptions.fullscreen.hidden)
            await this._fullEl.setJeepFullscreenVisibility({ visibility: 'visible' });
        return;
    }
    async _getCurrentSlideIndex() {
        let index = 0;
        this._slidesParams.position = this._slidesParams.axis === 'x' ? Number(this._slidesEl.scrollLeft) : Number(this._slidesEl.scrollTop);
        if (Math.abs(this._slidesParams.scrollLength / this._slidesParams.nbSlides - this._slidesParams.slideLength) < 2) {
            // slide constant width
            index = Math.round(this._slidesParams.position / this._slidesParams.slideLength);
        }
        return index;
    }
    async _goPrev() {
        this._slidesParams.curSlide -= 1;
        if (this._slidesParams.curSlide < 0)
            this._slidesParams.curSlide = 0;
        if (this.innerOptions.autoplay) {
            this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
        }
        else {
            this._setNavigationButtons();
            if (this.innerOptions.pagination)
                this._syncPagination();
        }
        await this.goToSlideIndex(this._slidesParams.curSlide);
    }
    async _goNext() {
        this._slidesParams.curSlide += 1;
        if (this._slidesParams.curSlide >= this._slidesParams.nbSlides)
            this._slidesParams.curSlide = this._slidesParams.nbSlides - 1;
        if (this.innerOptions.autoplay) {
            this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
        }
        else {
            this._setNavigationButtons();
            if (this.innerOptions.pagination)
                this._syncPagination();
        }
        await this.goToSlideIndex(this._slidesParams.curSlide);
    }
    async _setNavigationButtons() {
        if (this._navEl) {
            if (this._slidesParams.curSlide === 0)
                await this._navEl.setJeepNavigationPrevDisabled({ disabled: true });
            if (this._slidesParams.curSlide >= 1)
                await this._navEl.setJeepNavigationPrevDisabled({ disabled: false });
            if (this._slidesParams.curSlide === this._slidesParams.nbSlides - 1)
                await this._navEl.setJeepNavigationNextDisabled({ disabled: true });
            if (this._slidesParams.curSlide <= this._slidesParams.nbSlides - 2)
                await this._navEl.setJeepNavigationNextDisabled({ disabled: false });
        }
    }
    _syncPagination() {
        if (this._pagEl)
            this._pagEl.setJeepPaginationActiveIndex({ activeIndex: this._slidesParams.curSlide.toString() });
    }
    async _scrollToPosition(index) {
        if (index !== this._slidesParams.curSlide)
            this._slidesParams.curSlide = index;
        this._slidesParams.position = this._slidesParams.curSlide * this._slidesParams.slideLength;
        if (this._slidesParams.axis === 'x') {
            if (this._slidesEl)
                this._slidesEl.scrollTo(this._slidesParams.position, 0);
        }
        else {
            if (this._slidesEl)
                this._slidesEl.scrollTo(0, this._slidesParams.position);
        }
    }
    render() {
        return (core.h(core.Host, null, core.h("div", { class: "slides-container" }, core.h("div", { class: "slides-wrapper" }, core.h("div", { class: "slides" }, core.h("slot", { name: "slides" }))), !this.innerOptions.autoplay && this.innerOptions.navigation
            ? core.h("jeep-navigation", { id: "jeep-slides-navigation", name: this._navIcon })
            : null, !this.innerOptions.autoplay && this.innerOptions.pagination
            ? core.h("jeep-pagination", { id: `jeep-slides-pagination-${this._direction}`, ndisplay: this.innerOptions.pagination.bulletsDisplay ? this.innerOptions.pagination.bulletsDisplay : 5, clickable: this.innerOptions.pagination.clickable ? this.innerOptions.pagination.clickable : false, direction: this._direction })
            : null, this.innerOptions.autoplay
            ? core.h("jeep-play-controls", { id: "jeep-slides-autoplay", duration: this.innerOptions.autoplay.duration, nslides: this.nSlides })
            : null, this.innerOptions.fullscreen
            ? core.h("jeep-fullscreen", { id: "jeep-slides-fullscreen" })
            : null)));
    }
    get el() { return core.getElement(this); }
    static get watchers() { return {
        "options": ["parseOptionsProp"]
    }; }
    static get style() { return ":host{--slides-width:100vw;--slides-height:100vh;--slides-margin-top:var(--gslides-margin-top,0px);--slides-margin-left:var(--gslides-margin-left,0px);--slides-margin-right:var(--gslides-margin-right,0px);--slides-margin-bottom:var(--gslides-margin-bottom,0px);--slides-background-color:var(--gslides-background-color,transparent);--slides-pagination-bottom:var(--gslides-pagination-bottom,15px);--slides-pagination-right:var(--gslides-pagination-right,5px);--slides-bullet-active-diameter:var(--gslides-bullet-active-diameter,24px);--slides-bullet-diameter:var(--gslides-bullet-diameter,15px);--slides-bullet-background:var(--gslides-bullet-background,#000);--slides-bullet-opacity:var(--gslides-bullet-opacity,0.5);--slides-bullet-active-background:var(--gslides-bullet-active-background,#ff8000);--slides-bullet-active-opacity:var(--gslides-bullet-active-opacity,1);--slides-navigation-button-width:var(--gslides-navigation-button-width,50px);--slides-navigation-button-height:var(--gslides-navigation-button-height,50px);--slides-navigation-padding-horizontal:var(--gslides-navigation-padding-horizontal,10px);--slides-navigation-button-icon-color:var(--gslides-navigation-button-icon-color,#222428);--slides-slide-scroll-snap-align:var(--gslides-slide-scroll-snap-align,start);--slides-slide-background:var(--gslides-slide-background);--slides-slide-color:var(--gslides-slide-color);--slides-slide-title-font-size:var(--gslides-slide-title-font-size);--slides-slide-subtitle-font-size:var(--gslides-slide-subtitle-font-size);--slides-slide-content-top:var(--gslides-slide-content-top);--slides-slide-content-padding:var(--gslides-slide-content-padding);--slides-slide-content-font-size:var(--gslides-slide-content-font-size);--slides-slide-content-text-align:var(--gslides-slide-content-text-align);--slides-slide-header-top:var(--gslides-slide-header-top);--slides-playcontrols-bottom:var(--gslides-playcontrols-bottom,20px);--slides-playcontrols-button-width:var(--gslides-playcontrols-button-width,50px);--slides-playcontrols-button-height:var(--gslides-playcontrols-button-height,50px);--slides-fullscreen-top:var(--gslides-fullscreen-top,10px);--slides-fullscreen-left:var(--gslides-fullscreen-left,10px);--slides-fullscreen-button-width:var(--gslides-fullscreen-button-width,50px);--slides-fullscreen-button-height:var(--gslides-fullscreen-button-height,50px);--slides-fullscreen-button-icon-color:var(--gslides-fullscreen-button-icon-color,#222428)}.slides-container{position:absolute;top:0;bottom:0;right:0;left:0;overflow:hidden;background-color:var(--slides-background-color)}.slides-wrapper{margin-top:var(--slides-margin-top);margin-bottom:var(--slides-margin-bottom);margin-left:var(--slides-margin-left);margin-right:var(--slides-margin-right);position:relative;width:100%;height:100%}.slides,.slides-wrapper{-webkit-box-sizing:border-box;box-sizing:border-box;padding:0}.slides{margin:0;font-family:sans-serif;width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));height:calc(var(--slides-height) - var(--slides-margin-top) - var(--slides-margin-bottom));display:-ms-flexbox;display:flex;overflow:auto;scroll-snap-type:mandatory;-ms-scroll-snap-type:mandatory;-webkit-scroll-snap-type:mandatory;-webkit-scroll-snap-destination:0 0;-webkit-overflow-scrolling:touch}.slides.slides-horizontal{-webkit-scroll-snap-type:x mandatory;-ms-scroll-snap-type:x mandatory;scroll-snap-type:x mandatory;-webkit-scroll-snap-points-x:repeat(100%);scroll-snap-points-x:repeat(100%);-ms-scroll-snap-points-x:repeat(100%);overflow-x:scroll;-ms-flex-direction:row;flex-direction:row;overflow-y:hidden}.slides.slides-vertical{-webkit-scroll-snap-type:y mandatory;-ms-scroll-snap-type:y mandatory;scroll-snap-type:y mandatory;-webkit-scroll-snap-points-x:initial;scroll-snap-points-x:none;-ms-scroll-snap-points-x:initial;-webkit-scroll-snap-points-y:repeat(100%);scroll-snap-points-y:repeat(100%);-ms-scroll-snap-points-y:repeat(100%);overflow-y:scroll;-ms-flex-direction:column;flex-direction:column;overflow-x:hidden}#jeep-slides-navigation{--navigation-top:calc(var(--slides-height) / 2 + var(--slides-margin-top) / 2 - var(--slides-margin-bottom) / 2);--navigation-left:var(--slides-margin-left);--navigation-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));--navigation-button-width:var(--slides-navigation-button-width);--navigation-button-height:var(--slides-navigation-button-height);--navigation-padding-horizontal:var(--slides-navigation-padding-horizontal);--navigation-button-icon-color:var(--slides-navigation-button-icon-color)}#jeep-slides-pagination-horizontal{--pagination-top:calc(var(--slides-height) - var(--slides-margin-bottom) - var(--slides-pagination-bottom));--pagination-left:var(--slides-margin-left);--pagination-height:auto;--pagination-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right))}#jeep-slides-pagination-horizontal,#jeep-slides-pagination-vertical{--bullet-active-diameter:var(--slides-bullet-active-diameter);--bullet-diameter:var(--slides-bullet-diameter);--bullet-background:var(--slides-bullet-background);--bullet-opacity:var(--slides-bullet-opacity);--bullet-active-background:var(--slides-bullet-active-background);--bullet-active-opacity:var(--slides-bullet-active-opacity)}#jeep-slides-pagination-vertical{--pagination-top:var(--slides-margin-top);--pagination-left:calc(var(--slides-width) - var(--slides-pagination-right) - var(--slides-margin-right) - var(--slides-bullet-active-diameter));--pagination-width:calc(var(--slides-bullet-active-diameter));--pagination-height:calc(var(--slides-height) - var(--slides-margin-top) - var(--slides-margin-bottom));visibility:visible}#jeep-slides-autoplay{--playcontrols-top:calc(var(--slides-height) - var(--slides-margin-bottom) - var(--slides-playcontrols-bottom) - var(--slides-playcontrols-button-height));--playcontrols-left:var(--slides-margin-left);--playcontrols-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));--playcontrols-button-width:var(--slides-playcontrols-button-width);--playcontrols-button-height:var(--slides-playcontrols-button-height);visibility:visible}#jeep-slides-fullscreen{--fullscreen-top:calc(var(--slides-margin-top) + var(--slides-fullscreen-top));--fullscreen-left:calc(var(--slides-margin-left) + var(--slides-fullscreen-left));--fullscreen-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right) - var(--slides-fullscreen-left));--fullscreen-button-width:var(--slides-fullscreen-button-width);--fullscreen-button-height:var(--slides-fullscreen-button-height);--fullscreen-button-icon-color:var(--slides-fullscreen-button-icon-color);visibility:visible}jeep-slide{--slide-scroll-snap-align:var(--slides-slide-scroll-snap-align);--slide-min-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));--slide-height:calc(var(--slides-height) - var(--slides-margin-top) - var(--slides-margin-bottom))}"; }
};

exports.jeep_slides = JeepSlides;
