// @jeepq/core/dist/cjs/jeep-slide.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-ce02df39.js');
const common = require('./common-5e122d9c.js');

const JeepSlide = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        this.onSlideDidLoad = core.createEvent(this, "jeepSlideDidLoad", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************/
    parseTitleProp(newValue) {
        if (newValue)
            this.innerTitle = newValue;
    }
    parseSubTitleProp(newValue) {
        if (newValue)
            this.innerSubTitle = newValue;
    }
    async parseStyleProp(newValue) {
        if (newValue) {
            this._lcssvar = await common.getCssPropertyFromString(newValue, 'slide');
            this._setLocalCssVariables();
            this.innerStyle = newValue;
        }
    }
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    init() {
        return Promise.resolve(this._init());
    }
    /**
     * Set the slide component.
     */
    setSlide() {
        return Promise.resolve(this._setSlide());
    }
    async componentWillLoad() {
        await this.init();
    }
    async componentDidLoad() {
        await this.setSlide();
    }
    async _init() {
        this._element = this.el.shadowRoot;
        this.parseTitleProp(this.stitle ? this.stitle : null);
        this.parseSubTitleProp(this.subtitle ? this.subtitle : null);
        await this.parseStyleProp(this.cstyle ? this.cstyle : null);
        return;
    }
    _setLocalCssVariables() {
        if (this._lcssvar.alignItems)
            this.el.style.setProperty('--slide-align-items', this._lcssvar.alignItems);
        if (this._lcssvar.background)
            this.el.style.setProperty('--slide-background', this._lcssvar.background);
        if (this._lcssvar.color)
            this.el.style.setProperty('--slide-color', this._lcssvar.color);
        if (this._lcssvar.contentFontSize)
            this.el.style.setProperty('--slide-content-font-size', this._lcssvar.contentFontSize);
        if (this._lcssvar.contentPadding)
            this.el.style.setProperty('--slide-content-padding', this._lcssvar.contentPadding);
        if (this._lcssvar.contentTextAlign)
            this.el.style.setProperty('--slide-content-text-align', this._lcssvar.contentTextAlign);
        if (this._lcssvar.contentTop)
            this.el.style.setProperty('--slide-content-top', this._lcssvar.contentTop);
        if (this._lcssvar.display)
            this.el.style.setProperty('--slide-display', this._lcssvar.display);
        if (this._lcssvar.flexDirection)
            this.el.style.setProperty('--slide-flex-direction', this._lcssvar.flexDirection);
        if (this._lcssvar.fontSize)
            this.el.style.setProperty('--slide-font-size', this._lcssvar.fontSize);
        if (this._lcssvar.headerTop)
            this.el.style.setProperty('--slide-header-top', this._lcssvar.headerTop);
        if (this._lcssvar.height)
            this.el.style.setProperty('--slide-height', this._lcssvar.height);
        if (this._lcssvar.justifyContent)
            this.el.style.setProperty('--slide-justify-content', this._lcssvar.justifyContent);
        if (this._lcssvar.minWidth)
            this.el.style.setProperty('--slide-min-width', this._lcssvar.minWidth);
        if (this._lcssvar.scrollSnapAlign)
            this.el.style.setProperty('--slide-scroll-snap-align', this._lcssvar.scrollSnapAlign);
        if (this._lcssvar.subtitleFontSize)
            this.el.style.setProperty('--slide-subtitle-font-size', this._lcssvar.subtitleFontSize);
        if (this._lcssvar.textAlign)
            this.el.style.setProperty('--slide-text-align', this._lcssvar.textAlign);
        if (this._lcssvar.titleFontSize)
            this.el.style.setProperty('--slide-title-font-size', this._lcssvar.titleFontSize);
    }
    async _setSlide() {
        this.onSlideDidLoad.emit();
        return;
    }
    render() {
        return (core.h(core.Host, null, core.h("div", { class: "slide-container" }, core.h("div", { class: "slide-wrapper" }, core.h("div", { class: "slide-header" }, this.innerTitle != null
            ? core.h("div", { class: "slide-title" }, this.innerTitle)
            : null, this.innerSubTitle != null
            ? core.h("div", { class: "slide-subtitle" }, this.innerSubTitle)
            : null), core.h("div", { class: "slide-content" }, core.h("slot", null))))));
    }
    get el() { return core.getElement(this); }
    static get watchers() { return {
        "stitle": ["parseTitleProp"],
        "subtitle": ["parseSubTitleProp"],
        "cstyle": ["parseStyleProp"]
    }; }
    static get style() { return ":host{--slide-scroll-snap-align:start;--slide-min-width:100vw;--slide-height:100vh;--slide-background:var(--slides-slide-background,#fff);--slide-color:var(--slides-slide-color,#000);--slide-font-size:calc(1rem + 3vmin);--slide-display:flex;--slide-align-items:center;--slide-justify-content:center;--slide-text-align:center;--slide-flex-direction:column;--slide-title-font-size:var(--slides-slide-title-font-size,calc(1rem + 3vmin));--slide-subtitle-font-size:var(--slides-slide-subtitle-font-size,calc(1rem + 1.5vmin));--slide-content-top:var(--slides-slide-content-top,0px);--slide-content-padding:var(--slides-slide-content-padding,0);--slide-content-font-size:var(--slides-slide-content-font-size,calc(1rem + 0.5vmin));--slide-content-text-align:var(--slides-slide-content-text-align,center);--slide-header-top:var(--slides-slide-header-top,0px)}.slide-container{z-index:2;-webkit-box-sizing:border-box;box-sizing:border-box;scroll-snap-align:var(--slide-scroll-snap-align);-ms-scroll-snap-coordinate:0 0;scroll-snap-coordinate:0 0;-webkit-scroll-snap-coordinate:0 0;min-width:var(--slide-min-width);height:var(--slide-height);color:var(--slide-color);background:var(--slide-background)}.slide-container,.slide-wrapper{position:relative;margin:0 auto;-ms-flex-negative:0;flex-shrink:0}.slide-wrapper{-webkit-box-sizing:var(--slide-box-sizing);box-sizing:var(--slide-box-sizing);padding:0;display:var(--slide-display);-ms-flex-direction:var(--slide-flex-direction);flex-direction:var(--slide-flex-direction);-ms-flex-align:var(--slide-align-items);align-items:var(--slide-align-items);-ms-flex-pack:var(--slide-justify-content);justify-content:var(--slide-justify-content);width:100%;height:100%;font-size:var(--slide-font-size)}.slide-header{position:relative;top:var(--slide-header-top);text-align:center}.slide-title{font-size:var(--slide-title-font-size);font-weight:700}.slide-subtitle{text-align:center;font-size:var(--slide-subtitle-font-size)}.slide-content{position:relative;top:var(--slide-content-top);padding:var(--slide-content-padding);text-align:var(--slide-content-text-align);font-size:var(--slide-content-font-size)}.slide-image{width:auto;max-width:100vw;height:auto;max-height:100vh}"; }
};

exports.jeep_slide = JeepSlide;
