// @jeepq/core/dist/esm-es5/jeep-slide.entry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-f19ae6a4.js';
import { a as getCssPropertyFromString } from './common-61073bba.js';
var JeepSlide = /** @class */ (function () {
    function class_1(hostRef) {
        registerInstance(this, hostRef);
        this.onSlideDidLoad = createEvent(this, "jeepSlideDidLoad", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************/
    class_1.prototype.parseTitleProp = function (newValue) {
        if (newValue)
            this.innerTitle = newValue;
    };
    class_1.prototype.parseSubTitleProp = function (newValue) {
        if (newValue)
            this.innerSubTitle = newValue;
    };
    class_1.prototype.parseStyleProp = function (newValue) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!newValue) return [3 /*break*/, 2];
                        _a = this;
                        return [4 /*yield*/, getCssPropertyFromString(newValue, 'slide')];
                    case 1:
                        _a._lcssvar = _b.sent();
                        this._setLocalCssVariables();
                        this.innerStyle = newValue;
                        _b.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_1.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    /**
     * Set the slide component.
     */
    class_1.prototype.setSlide = function () {
        return Promise.resolve(this._setSlide());
    };
    class_1.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.setSlide()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._element = this.el.shadowRoot;
                        this.parseTitleProp(this.stitle ? this.stitle : null);
                        this.parseSubTitleProp(this.subtitle ? this.subtitle : null);
                        return [4 /*yield*/, this.parseStyleProp(this.cstyle ? this.cstyle : null)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._setLocalCssVariables = function () {
        if (this._lcssvar.alignItems)
            this.el.style.setProperty('--slide-align-items', this._lcssvar.alignItems);
        if (this._lcssvar.background)
            this.el.style.setProperty('--slide-background', this._lcssvar.background);
        if (this._lcssvar.color)
            this.el.style.setProperty('--slide-color', this._lcssvar.color);
        if (this._lcssvar.contentFontSize)
            this.el.style.setProperty('--slide-content-font-size', this._lcssvar.contentFontSize);
        if (this._lcssvar.contentPadding)
            this.el.style.setProperty('--slide-content-padding', this._lcssvar.contentPadding);
        if (this._lcssvar.contentTextAlign)
            this.el.style.setProperty('--slide-content-text-align', this._lcssvar.contentTextAlign);
        if (this._lcssvar.contentTop)
            this.el.style.setProperty('--slide-content-top', this._lcssvar.contentTop);
        if (this._lcssvar.display)
            this.el.style.setProperty('--slide-display', this._lcssvar.display);
        if (this._lcssvar.flexDirection)
            this.el.style.setProperty('--slide-flex-direction', this._lcssvar.flexDirection);
        if (this._lcssvar.fontSize)
            this.el.style.setProperty('--slide-font-size', this._lcssvar.fontSize);
        if (this._lcssvar.headerTop)
            this.el.style.setProperty('--slide-header-top', this._lcssvar.headerTop);
        if (this._lcssvar.height)
            this.el.style.setProperty('--slide-height', this._lcssvar.height);
        if (this._lcssvar.justifyContent)
            this.el.style.setProperty('--slide-justify-content', this._lcssvar.justifyContent);
        if (this._lcssvar.minWidth)
            this.el.style.setProperty('--slide-min-width', this._lcssvar.minWidth);
        if (this._lcssvar.scrollSnapAlign)
            this.el.style.setProperty('--slide-scroll-snap-align', this._lcssvar.scrollSnapAlign);
        if (this._lcssvar.subtitleFontSize)
            this.el.style.setProperty('--slide-subtitle-font-size', this._lcssvar.subtitleFontSize);
        if (this._lcssvar.textAlign)
            this.el.style.setProperty('--slide-text-align', this._lcssvar.textAlign);
        if (this._lcssvar.titleFontSize)
            this.el.style.setProperty('--slide-title-font-size', this._lcssvar.titleFontSize);
    };
    class_1.prototype._setSlide = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.onSlideDidLoad.emit();
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype.render = function () {
        return (h(Host, null, h("div", { class: "slide-container" }, h("div", { class: "slide-wrapper" }, h("div", { class: "slide-header" }, this.innerTitle != null
            ? h("div", { class: "slide-title" }, this.innerTitle)
            : null, this.innerSubTitle != null
            ? h("div", { class: "slide-subtitle" }, this.innerSubTitle)
            : null), h("div", { class: "slide-content" }, h("slot", null))))));
    };
    Object.defineProperty(class_1.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "watchers", {
        get: function () {
            return {
                "stitle": ["parseTitleProp"],
                "subtitle": ["parseSubTitleProp"],
                "cstyle": ["parseStyleProp"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "style", {
        get: function () { return ":host{--slide-scroll-snap-align:start;--slide-min-width:100vw;--slide-height:100vh;--slide-background:var(--slides-slide-background,#fff);--slide-color:var(--slides-slide-color,#000);--slide-font-size:calc(1rem + 3vmin);--slide-display:flex;--slide-align-items:center;--slide-justify-content:center;--slide-text-align:center;--slide-flex-direction:column;--slide-title-font-size:var(--slides-slide-title-font-size,calc(1rem + 3vmin));--slide-subtitle-font-size:var(--slides-slide-subtitle-font-size,calc(1rem + 1.5vmin));--slide-content-top:var(--slides-slide-content-top,0px);--slide-content-padding:var(--slides-slide-content-padding,0);--slide-content-font-size:var(--slides-slide-content-font-size,calc(1rem + 0.5vmin));--slide-content-text-align:var(--slides-slide-content-text-align,center);--slide-header-top:var(--slides-slide-header-top,0px)}.slide-container{z-index:2;-webkit-box-sizing:border-box;box-sizing:border-box;scroll-snap-align:var(--slide-scroll-snap-align);-ms-scroll-snap-coordinate:0 0;scroll-snap-coordinate:0 0;-webkit-scroll-snap-coordinate:0 0;min-width:var(--slide-min-width);height:var(--slide-height);color:var(--slide-color);background:var(--slide-background)}.slide-container,.slide-wrapper{position:relative;margin:0 auto;-ms-flex-negative:0;flex-shrink:0}.slide-wrapper{-webkit-box-sizing:var(--slide-box-sizing);box-sizing:var(--slide-box-sizing);padding:0;display:var(--slide-display);-ms-flex-direction:var(--slide-flex-direction);flex-direction:var(--slide-flex-direction);-ms-flex-align:var(--slide-align-items);align-items:var(--slide-align-items);-ms-flex-pack:var(--slide-justify-content);justify-content:var(--slide-justify-content);width:100%;height:100%;font-size:var(--slide-font-size)}.slide-header{position:relative;top:var(--slide-header-top);text-align:center}.slide-title{font-size:var(--slide-title-font-size);font-weight:700}.slide-subtitle{text-align:center;font-size:var(--slide-subtitle-font-size)}.slide-content{position:relative;top:var(--slide-content-top);padding:var(--slide-content-padding);text-align:var(--slide-content-text-align);font-size:var(--slide-content-font-size)}.slide-image{width:auto;max-width:100vw;height:auto;max-height:100vh}"; },
        enumerable: true,
        configurable: true
    });
    return class_1;
}());
export { JeepSlide as jeep_slide };
