// @jeepq/core/dist/esm-es5/jeep-slides.entry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-f19ae6a4.js';
import { g as getBoundingClientRect } from './common-61073bba.js';
var JeepSlides = /** @class */ (function () {
    function class_1(hostRef) {
        registerInstance(this, hostRef);
        this.fullScreen = false;
        this._slidesParams = {};
        this._ticking = false;
        this._oriMargin = {};
        this._touchStart = false;
        this._touchMove = false;
        this.onSlidesHeaderVisibility = createEvent(this, "jeepSlidesHeaderVisibility", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    class_1.prototype.parseOptionsProp = function (newValue) {
        if (newValue)
            this.innerOptions = JSON.parse(newValue);
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    class_1.prototype.handleWindowResize = function () {
        this._windowResize();
    };
    class_1.prototype.handleJeepNavigationPrev = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.goToPreviousSlide()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleJeepNavigationNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.goToNextSlide()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleJeepPaginationIndex = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(event.detail && event.detail.index)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.goToSlideIndex(Number(event.detail.index))];
                    case 1:
                        _a.sent();
                        if (this.innerOptions.navigation)
                            this._setNavigationButtons();
                        return [4 /*yield*/, this._pagEl.setJeepPaginationActiveIndex({ activeIndex: this._slidesParams.curSlide.toString() })];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleJeepPlayControlsCurrentIndex = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(event.detail && event.detail.index)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.goToSlideIndex(Number(event.detail.index))];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleJeepFullscreenRequest = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._setProperties(false);
                        return [4 /*yield*/, this._fullEl.fullscreenRequest(document.documentElement)];
                    case 1:
                        _a.sent();
                        this.fullScreen = true;
                        this._windowResize();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleJeepFullscreenExit = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._fullEl.fullscreenExit()];
                    case 1:
                        _a.sent();
                        this._setProperties(true);
                        this.fullScreen = false;
                        this._windowResize();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleJeepFullscreenChange = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._setProperties(true);
                this.fullScreen = false;
                this._windowResize();
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype.handleKeydown = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var key, keycode, pressKey, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        event.preventDefault();
                        key = event.key ? event.key : null;
                        keycode = event.keyCode || event.which ? event.keyCode || event.which : null;
                        pressKey = key !== null ? key : keycode.toString();
                        _a = pressKey;
                        switch (_a) {
                            case 'ArrowLeft': return [3 /*break*/, 1];
                            case 'ArrowUp': return [3 /*break*/, 4];
                            case 'ArrowRight': return [3 /*break*/, 7];
                            case 'ArrowDown': return [3 /*break*/, 10];
                        }
                        return [3 /*break*/, 13];
                    case 1:
                        if (!(this._direction === 'horizontal')) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.goToPreviousSlide()];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3: return [3 /*break*/, 13];
                    case 4:
                        if (!(this._direction === 'vertical')) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.goToPreviousSlide()];
                    case 5:
                        _b.sent();
                        _b.label = 6;
                    case 6: return [3 /*break*/, 13];
                    case 7:
                        if (!(this._direction === 'horizontal')) return [3 /*break*/, 9];
                        return [4 /*yield*/, this.goToNextSlide()];
                    case 8:
                        _b.sent();
                        _b.label = 9;
                    case 9: return [3 /*break*/, 13];
                    case 10:
                        if (!(this._direction === 'vertical')) return [3 /*break*/, 12];
                        return [4 /*yield*/, this.goToNextSlide()];
                    case 11:
                        _b.sent();
                        _b.label = 12;
                    case 12: return [3 /*break*/, 13];
                    case 13: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.handleTouchstart = function () {
        this._touchStart = true;
    };
    class_1.prototype.handleTouchmove = function () {
        this._touchMove = true;
    };
    class_1.prototype.handleTouchend = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(!this._touchStart || !this._touchMove)) return [3 /*break*/, 2];
                        _a = this;
                        return [4 /*yield*/, this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500)];
                    case 1:
                        _a._initial = _b.sent();
                        _b.label = 2;
                    case 2:
                        this._touchStart = false;
                        this._touchMove = false;
                        return [2 /*return*/];
                }
            });
        });
    };
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_1.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    /**
     * Set the slides.
     */
    class_1.prototype.setSlides = function () {
        return Promise.resolve(this._setSlides());
    };
    /**
     * Get Active Slide Index.
     */
    class_1.prototype.getActiveSlideIndex = function () {
        return __awaiter(this, void 0, void 0, function () {
            var curSlide;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._getCurrentSlideIndex()];
                    case 1:
                        curSlide = _a.sent();
                        return [2 /*return*/, Promise.resolve(curSlide)];
                }
            });
        });
    };
    /**
     *  Go Previous Slide
     */
    class_1.prototype.goToPreviousSlide = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._goPrev()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     *  Go Next Slide
     */
    class_1.prototype.goToNextSlide = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._goNext()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     *  Go To Slide with given index
     */
    class_1.prototype.goToSlideIndex = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._scrollToPosition(index)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    class_1.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.setSlides()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            var defOpt;
            return __generator(this, function (_a) {
                this._element = this.el.shadowRoot;
                this._slidesNumber = 0;
                defOpt = { direction: "horizontal" };
                this.parseOptionsProp(this.options ? this.options : JSON.stringify(defOpt));
                this._direction = this.innerOptions.direction ? this.innerOptions.direction : "horizontal";
                this._navIcon = this.innerOptions.navigation && this.innerOptions.navigation.icon ? this.innerOptions.navigation.icon : "arrow-circle";
                this._oriMargin.top = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-top');
                this._oriMargin.left = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-left');
                this._oriMargin.bottom = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-bottom');
                this._oriMargin.right = window.getComputedStyle(this.el).getPropertyValue('--slides-margin-right');
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._setSlides = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, i, div, i, pagId, _b;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        this._containerEl = this._element.querySelector('.slides-container');
                        this._wrapperEl = this._containerEl.querySelector('.slides-wrapper');
                        this._slidesEl = this._wrapperEl.querySelector('.slides');
                        this._slidesEl.classList.add("slides-" + this._direction);
                        this._slotEl = this._slidesEl.querySelector('slot');
                        if (!(this._slotEl !== null)) return [3 /*break*/, 8];
                        _a = this;
                        return [4 /*yield*/, this._getElementsInSlot()];
                    case 1:
                        _a._arrSlidesEl = _c.sent();
                        if (!(this._arrSlidesEl !== null && this._arrSlidesEl.length > 0)) return [3 /*break*/, 8];
                        //<-- required to be added as the scroll horizontal does not work with slot element
                        // loop through the slot element , add the elements to the render
                        // store elements which are in slot
                        this._slidesNumber = this._arrSlidesEl.length;
                        for (i = 0; i < this._slidesNumber; i++) {
                            this._slidesEl.append(this._arrSlidesEl[i]);
                        }
                        div = Array.from(this.el.querySelectorAll('div'));
                        if (div !== null) {
                            for (i = 0; i < div.length; i++) {
                                if (div[i].slot === "slides")
                                    this.el.removeChild(div[i]);
                            }
                            this._slidesEl.removeChild(this._slotEl);
                        }
                        //-->
                        if (this.innerOptions.fullscreen)
                            this._fullEl = this._containerEl.querySelector('#jeep-slides-fullscreen');
                        if (!this.innerOptions.autoplay && this.innerOptions.navigation)
                            this._navEl = this._containerEl.querySelector('#jeep-slides-navigation');
                        if (!(!this.innerOptions.autoplay && this.innerOptions.pagination)) return [3 /*break*/, 3];
                        pagId = "#jeep-slides-pagination-" + this.innerOptions.direction;
                        this._pagEl = this._containerEl.querySelector(pagId);
                        return [4 /*yield*/, this._pagEl.setJeepPaginationSlidesNumber({ slides: this._arrSlidesEl.length })];
                    case 2:
                        _c.sent();
                        _c.label = 3;
                    case 3:
                        if (this.innerOptions.autoplay)
                            this._playEl = this._containerEl.querySelector('#jeep-slides-autoplay');
                        // add scroll event
                        this._slidesEl.addEventListener('scroll', function () {
                            var prevIndex = _this._slidesParams.curSlide;
                            if (!_this._ticking) {
                                setTimeout(function () { return __awaiter(_this, void 0, void 0, function () {
                                    var _a;
                                    return __generator(this, function (_b) {
                                        switch (_b.label) {
                                            case 0:
                                                _a = this._slidesParams;
                                                return [4 /*yield*/, this.getActiveSlideIndex()];
                                            case 1:
                                                _a.curSlide = _b.sent();
                                                this._slidesParams.position = this._slidesParams.axis === 'x' ? Number(this._slidesEl.scrollLeft) : Number(this._slidesEl.scrollTop);
                                                if (this.innerOptions.navigation && this._slidesParams.curSlide !== prevIndex)
                                                    this._setNavigationButtons();
                                                if (this.innerOptions.pagination && this._slidesParams.curSlide !== prevIndex)
                                                    this._syncPagination();
                                                if (this.innerOptions.autoplay && this._slidesParams.curSlide !== prevIndex)
                                                    this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
                                                this._ticking = false;
                                                return [2 /*return*/];
                                        }
                                    });
                                }); }, 200);
                                _this._ticking = true;
                            }
                        }, false);
                        // add click event
                        this._slidesEl.addEventListener('click', function () { return __awaiter(_this, void 0, void 0, function () {
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        _a = this;
                                        return [4 /*yield*/, this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500)];
                                    case 1:
                                        _a._initial = _b.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); }, false);
                        // fill the slidesParams
                        this._slidesParams.axis = this._direction === 'vertical' ? 'y' : 'x';
                        this._slidesParams.curSlide = 0;
                        this._slidesParams.nbSlides = this._arrSlidesEl.length;
                        return [4 /*yield*/, this._initScrollParams()];
                    case 4:
                        _c.sent();
                        if (!(!this.innerOptions.autoplay && (this.innerOptions.navigation && this._slidesParams.nbSlides > 1))) return [3 /*break*/, 6];
                        return [4 /*yield*/, this._navEl.setJeepNavigationNextDisabled({ disabled: false })];
                    case 5:
                        _c.sent();
                        _c.label = 6;
                    case 6:
                        _b = this;
                        return [4 /*yield*/, this._doHide(this.innerOptions.duration ? this.innerOptions.duration : 2500)];
                    case 7:
                        _b._initial = _c.sent();
                        this.nSlides = this._arrSlidesEl.length;
                        _c.label = 8;
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._setProperties = function (back) {
        if (!back) {
            this.el.style.setProperty('--slides-margin-top', '0px');
            this.el.style.setProperty('--slides-margin-bottom', '0px');
            this.el.style.setProperty('--slides-margin-left', '0px');
            this.el.style.setProperty('--slides-margin-right', '0px');
        }
        else {
            this.el.style.setProperty('--slides-margin-top', this._oriMargin.top);
            this.el.style.setProperty('--slides-margin-bottom', this._oriMargin.bottom);
            this.el.style.setProperty('--slides-margin-left', this._oriMargin.left);
            this.el.style.setProperty('--slides-margin-right', this._oriMargin.right);
        }
    };
    class_1.prototype._windowResize = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.el.style.setProperty('--slides-width', window.innerWidth + "px");
                        this.el.style.setProperty('--slides-height', window.innerHeight + "px");
                        return [4 /*yield*/, this._initScrollParams()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.goToSlideIndex(this._slidesParams.curSlide)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._getElementsInSlot = function () {
        return __awaiter(this, void 0, void 0, function () {
            var elems, nodes, cmpSlides;
            return __generator(this, function (_a) {
                elems = null;
                nodes = this._slotEl.assignedNodes();
                cmpSlides = nodes.filter(function (item) {
                    return item.slot === 'slides' ? item : null;
                });
                if (cmpSlides !== null && cmpSlides.length > 0) {
                    elems = Array.from(cmpSlides[0].querySelectorAll('jeep-slide'));
                }
                return [2 /*return*/, elems];
            });
        });
    };
    class_1.prototype._initScrollParams = function () {
        return __awaiter(this, void 0, void 0, function () {
            var rect;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, getBoundingClientRect(this._arrSlidesEl[0], 100)];
                    case 1:
                        rect = _a.sent();
                        this._slidesParams.slideLength = this._slidesParams.axis === 'x' ? rect.width : rect.height;
                        this._slidesParams.scrollLength = this._slidesParams.axis === 'x' ? this._slidesEl.scrollWidth :
                            this._slidesEl.scrollHeight;
                        this._slidesParams.position = this._slidesParams.curSlide * this._slidesParams.slideLength;
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._doHide = function (duration) {
        return __awaiter(this, void 0, void 0, function () {
            var initial;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        clearTimeout(this._initial);
                        return [4 /*yield*/, this._doVisible()];
                    case 1:
                        _a.sent();
                        initial = setTimeout(function () { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        this.onSlidesHeaderVisibility.emit({ visibility: 'hidden' });
                                        if (!(this._navEl && this.innerOptions.navigation.hidden)) return [3 /*break*/, 2];
                                        return [4 /*yield*/, this._navEl.setJeepNavigationVisibility({ visibility: 'hidden' })];
                                    case 1:
                                        _a.sent();
                                        _a.label = 2;
                                    case 2:
                                        if (!(this._pagEl && this.innerOptions.pagination.hidden)) return [3 /*break*/, 4];
                                        return [4 /*yield*/, this._pagEl.setJeepPaginationVisibility({ visibility: 'hidden' })];
                                    case 3:
                                        _a.sent();
                                        _a.label = 4;
                                    case 4:
                                        if (!(this._playEl && this.innerOptions.autoplay.hidden)) return [3 /*break*/, 6];
                                        return [4 /*yield*/, this._playEl.setJeepPlayControlsVisibility({ visibility: 'hidden' })];
                                    case 5:
                                        _a.sent();
                                        _a.label = 6;
                                    case 6:
                                        if (!(this._fullEl && this.innerOptions.fullscreen.hidden)) return [3 /*break*/, 8];
                                        return [4 /*yield*/, this._fullEl.setJeepFullscreenVisibility({ visibility: 'hidden' })];
                                    case 7:
                                        _a.sent();
                                        _a.label = 8;
                                    case 8: return [2 /*return*/];
                                }
                            });
                        }); }, duration);
                        return [2 /*return*/, initial];
                }
            });
        });
    };
    class_1.prototype._doVisible = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this._slidesParams.curSlide === 0 || this._slidesParams.curSlide === this._slidesNumber - 1)
                            this.onSlidesHeaderVisibility.emit({ visibility: 'visible' });
                        if (!(this._navEl && this.innerOptions.navigation.hidden)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this._navEl.setJeepNavigationVisibility({ visibility: 'visible' })];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!(this._pagEl && this.innerOptions.pagination.hidden)) return [3 /*break*/, 4];
                        return [4 /*yield*/, this._pagEl.setJeepPaginationVisibility({ visibility: 'visible' })];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        if (!(this._playEl && this.innerOptions.autoplay.hidden)) return [3 /*break*/, 6];
                        return [4 /*yield*/, this._playEl.setJeepPlayControlsVisibility({ visibility: 'visible' })];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6:
                        if (!(this._fullEl && this.innerOptions.fullscreen.hidden)) return [3 /*break*/, 8];
                        return [4 /*yield*/, this._fullEl.setJeepFullscreenVisibility({ visibility: 'visible' })];
                    case 7:
                        _a.sent();
                        _a.label = 8;
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._getCurrentSlideIndex = function () {
        return __awaiter(this, void 0, void 0, function () {
            var index;
            return __generator(this, function (_a) {
                index = 0;
                this._slidesParams.position = this._slidesParams.axis === 'x' ? Number(this._slidesEl.scrollLeft) : Number(this._slidesEl.scrollTop);
                if (Math.abs(this._slidesParams.scrollLength / this._slidesParams.nbSlides - this._slidesParams.slideLength) < 2) {
                    // slide constant width
                    index = Math.round(this._slidesParams.position / this._slidesParams.slideLength);
                }
                return [2 /*return*/, index];
            });
        });
    };
    class_1.prototype._goPrev = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._slidesParams.curSlide -= 1;
                        if (this._slidesParams.curSlide < 0)
                            this._slidesParams.curSlide = 0;
                        if (this.innerOptions.autoplay) {
                            this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
                        }
                        else {
                            this._setNavigationButtons();
                            if (this.innerOptions.pagination)
                                this._syncPagination();
                        }
                        return [4 /*yield*/, this.goToSlideIndex(this._slidesParams.curSlide)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._goNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._slidesParams.curSlide += 1;
                        if (this._slidesParams.curSlide >= this._slidesParams.nbSlides)
                            this._slidesParams.curSlide = this._slidesParams.nbSlides - 1;
                        if (this.innerOptions.autoplay) {
                            this._playEl.setActiveIndexAndPlay(this._slidesParams.curSlide);
                        }
                        else {
                            this._setNavigationButtons();
                            if (this.innerOptions.pagination)
                                this._syncPagination();
                        }
                        return [4 /*yield*/, this.goToSlideIndex(this._slidesParams.curSlide)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._setNavigationButtons = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this._navEl) return [3 /*break*/, 8];
                        if (!(this._slidesParams.curSlide === 0)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this._navEl.setJeepNavigationPrevDisabled({ disabled: true })];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!(this._slidesParams.curSlide >= 1)) return [3 /*break*/, 4];
                        return [4 /*yield*/, this._navEl.setJeepNavigationPrevDisabled({ disabled: false })];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        if (!(this._slidesParams.curSlide === this._slidesParams.nbSlides - 1)) return [3 /*break*/, 6];
                        return [4 /*yield*/, this._navEl.setJeepNavigationNextDisabled({ disabled: true })];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6:
                        if (!(this._slidesParams.curSlide <= this._slidesParams.nbSlides - 2)) return [3 /*break*/, 8];
                        return [4 /*yield*/, this._navEl.setJeepNavigationNextDisabled({ disabled: false })];
                    case 7:
                        _a.sent();
                        _a.label = 8;
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._syncPagination = function () {
        if (this._pagEl)
            this._pagEl.setJeepPaginationActiveIndex({ activeIndex: this._slidesParams.curSlide.toString() });
    };
    class_1.prototype._scrollToPosition = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (index !== this._slidesParams.curSlide)
                    this._slidesParams.curSlide = index;
                this._slidesParams.position = this._slidesParams.curSlide * this._slidesParams.slideLength;
                if (this._slidesParams.axis === 'x') {
                    if (this._slidesEl)
                        this._slidesEl.scrollTo(this._slidesParams.position, 0);
                }
                else {
                    if (this._slidesEl)
                        this._slidesEl.scrollTo(0, this._slidesParams.position);
                }
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype.render = function () {
        return (h(Host, null, h("div", { class: "slides-container" }, h("div", { class: "slides-wrapper" }, h("div", { class: "slides" }, h("slot", { name: "slides" }))), !this.innerOptions.autoplay && this.innerOptions.navigation
            ? h("jeep-navigation", { id: "jeep-slides-navigation", name: this._navIcon })
            : null, !this.innerOptions.autoplay && this.innerOptions.pagination
            ? h("jeep-pagination", { id: "jeep-slides-pagination-" + this._direction, ndisplay: this.innerOptions.pagination.bulletsDisplay ? this.innerOptions.pagination.bulletsDisplay : 5, clickable: this.innerOptions.pagination.clickable ? this.innerOptions.pagination.clickable : false, direction: this._direction })
            : null, this.innerOptions.autoplay
            ? h("jeep-play-controls", { id: "jeep-slides-autoplay", duration: this.innerOptions.autoplay.duration, nslides: this.nSlides })
            : null, this.innerOptions.fullscreen
            ? h("jeep-fullscreen", { id: "jeep-slides-fullscreen" })
            : null)));
    };
    Object.defineProperty(class_1.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "watchers", {
        get: function () {
            return {
                "options": ["parseOptionsProp"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "style", {
        get: function () { return ":host{--slides-width:100vw;--slides-height:100vh;--slides-margin-top:var(--gslides-margin-top,0px);--slides-margin-left:var(--gslides-margin-left,0px);--slides-margin-right:var(--gslides-margin-right,0px);--slides-margin-bottom:var(--gslides-margin-bottom,0px);--slides-background-color:var(--gslides-background-color,transparent);--slides-pagination-bottom:var(--gslides-pagination-bottom,15px);--slides-pagination-right:var(--gslides-pagination-right,5px);--slides-bullet-active-diameter:var(--gslides-bullet-active-diameter,24px);--slides-bullet-diameter:var(--gslides-bullet-diameter,15px);--slides-bullet-background:var(--gslides-bullet-background,#000);--slides-bullet-opacity:var(--gslides-bullet-opacity,0.5);--slides-bullet-active-background:var(--gslides-bullet-active-background,#ff8000);--slides-bullet-active-opacity:var(--gslides-bullet-active-opacity,1);--slides-navigation-button-width:var(--gslides-navigation-button-width,50px);--slides-navigation-button-height:var(--gslides-navigation-button-height,50px);--slides-navigation-padding-horizontal:var(--gslides-navigation-padding-horizontal,10px);--slides-navigation-button-icon-color:var(--gslides-navigation-button-icon-color,#222428);--slides-slide-scroll-snap-align:var(--gslides-slide-scroll-snap-align,start);--slides-slide-background:var(--gslides-slide-background);--slides-slide-color:var(--gslides-slide-color);--slides-slide-title-font-size:var(--gslides-slide-title-font-size);--slides-slide-subtitle-font-size:var(--gslides-slide-subtitle-font-size);--slides-slide-content-top:var(--gslides-slide-content-top);--slides-slide-content-padding:var(--gslides-slide-content-padding);--slides-slide-content-font-size:var(--gslides-slide-content-font-size);--slides-slide-content-text-align:var(--gslides-slide-content-text-align);--slides-slide-header-top:var(--gslides-slide-header-top);--slides-playcontrols-bottom:var(--gslides-playcontrols-bottom,20px);--slides-playcontrols-button-width:var(--gslides-playcontrols-button-width,50px);--slides-playcontrols-button-height:var(--gslides-playcontrols-button-height,50px);--slides-fullscreen-top:var(--gslides-fullscreen-top,10px);--slides-fullscreen-left:var(--gslides-fullscreen-left,10px);--slides-fullscreen-button-width:var(--gslides-fullscreen-button-width,50px);--slides-fullscreen-button-height:var(--gslides-fullscreen-button-height,50px);--slides-fullscreen-button-icon-color:var(--gslides-fullscreen-button-icon-color,#222428)}.slides-container{position:absolute;top:0;bottom:0;right:0;left:0;overflow:hidden;background-color:var(--slides-background-color)}.slides-wrapper{margin-top:var(--slides-margin-top);margin-bottom:var(--slides-margin-bottom);margin-left:var(--slides-margin-left);margin-right:var(--slides-margin-right);position:relative;width:100%;height:100%}.slides,.slides-wrapper{-webkit-box-sizing:border-box;box-sizing:border-box;padding:0}.slides{margin:0;font-family:sans-serif;width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));height:calc(var(--slides-height) - var(--slides-margin-top) - var(--slides-margin-bottom));display:-ms-flexbox;display:flex;overflow:auto;scroll-snap-type:mandatory;-ms-scroll-snap-type:mandatory;-webkit-scroll-snap-type:mandatory;-webkit-scroll-snap-destination:0 0;-webkit-overflow-scrolling:touch}.slides.slides-horizontal{-webkit-scroll-snap-type:x mandatory;-ms-scroll-snap-type:x mandatory;scroll-snap-type:x mandatory;-webkit-scroll-snap-points-x:repeat(100%);scroll-snap-points-x:repeat(100%);-ms-scroll-snap-points-x:repeat(100%);overflow-x:scroll;-ms-flex-direction:row;flex-direction:row;overflow-y:hidden}.slides.slides-vertical{-webkit-scroll-snap-type:y mandatory;-ms-scroll-snap-type:y mandatory;scroll-snap-type:y mandatory;-webkit-scroll-snap-points-x:initial;scroll-snap-points-x:none;-ms-scroll-snap-points-x:initial;-webkit-scroll-snap-points-y:repeat(100%);scroll-snap-points-y:repeat(100%);-ms-scroll-snap-points-y:repeat(100%);overflow-y:scroll;-ms-flex-direction:column;flex-direction:column;overflow-x:hidden}#jeep-slides-navigation{--navigation-top:calc(var(--slides-height) / 2 + var(--slides-margin-top) / 2 - var(--slides-margin-bottom) / 2);--navigation-left:var(--slides-margin-left);--navigation-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));--navigation-button-width:var(--slides-navigation-button-width);--navigation-button-height:var(--slides-navigation-button-height);--navigation-padding-horizontal:var(--slides-navigation-padding-horizontal);--navigation-button-icon-color:var(--slides-navigation-button-icon-color)}#jeep-slides-pagination-horizontal{--pagination-top:calc(var(--slides-height) - var(--slides-margin-bottom) - var(--slides-pagination-bottom));--pagination-left:var(--slides-margin-left);--pagination-height:auto;--pagination-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right))}#jeep-slides-pagination-horizontal,#jeep-slides-pagination-vertical{--bullet-active-diameter:var(--slides-bullet-active-diameter);--bullet-diameter:var(--slides-bullet-diameter);--bullet-background:var(--slides-bullet-background);--bullet-opacity:var(--slides-bullet-opacity);--bullet-active-background:var(--slides-bullet-active-background);--bullet-active-opacity:var(--slides-bullet-active-opacity)}#jeep-slides-pagination-vertical{--pagination-top:var(--slides-margin-top);--pagination-left:calc(var(--slides-width) - var(--slides-pagination-right) - var(--slides-margin-right) - var(--slides-bullet-active-diameter));--pagination-width:calc(var(--slides-bullet-active-diameter));--pagination-height:calc(var(--slides-height) - var(--slides-margin-top) - var(--slides-margin-bottom));visibility:visible}#jeep-slides-autoplay{--playcontrols-top:calc(var(--slides-height) - var(--slides-margin-bottom) - var(--slides-playcontrols-bottom) - var(--slides-playcontrols-button-height));--playcontrols-left:var(--slides-margin-left);--playcontrols-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));--playcontrols-button-width:var(--slides-playcontrols-button-width);--playcontrols-button-height:var(--slides-playcontrols-button-height);visibility:visible}#jeep-slides-fullscreen{--fullscreen-top:calc(var(--slides-margin-top) + var(--slides-fullscreen-top));--fullscreen-left:calc(var(--slides-margin-left) + var(--slides-fullscreen-left));--fullscreen-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right) - var(--slides-fullscreen-left));--fullscreen-button-width:var(--slides-fullscreen-button-width);--fullscreen-button-height:var(--slides-fullscreen-button-height);--fullscreen-button-icon-color:var(--slides-fullscreen-button-icon-color);visibility:visible}jeep-slide{--slide-scroll-snap-align:var(--slides-slide-scroll-snap-align);--slide-min-width:calc(var(--slides-width) - var(--slides-margin-left) - var(--slides-margin-right));--slide-height:calc(var(--slides-height) - var(--slides-margin-top) - var(--slides-margin-bottom))}"; },
        enumerable: true,
        configurable: true
    });
    return class_1;
}());
export { JeepSlides as jeep_slides };
