// @jeepq/core/dist/esm-es5/jeep-html-toprint.entry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-f19ae6a4.js';
var JeepHtmlToprint = /** @class */ (function () {
    function class_1(hostRef) {
        registerInstance(this, hostRef);
        this.innerSlotStyle = null;
        this._error = null;
        this.printPage = createEvent(this, "jeepHtmlToPrint", 7);
        this.printPageReady = createEvent(this, "jeepHtmlToPrintReady", 7);
        this.printPageCompleted = createEvent(this, "jeepHtmlToPrintCompleted", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    class_1.prototype.parseSlotStyleProp = function (newValue) {
        if (newValue)
            this.innerSlotStyle = newValue ? newValue : null;
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    class_1.prototype.printDocHandler = function () {
        this.window.print();
        this.printPageCompleted.emit();
    };
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Method initialize
     */
    class_1.prototype.init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._init()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    class_1.prototype.load = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._load()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    class_1.prototype.emitPrint = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.printPage.emit();
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype.getSlotStyle = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.innerSlotStyle];
            });
        });
    };
    //*******************************
    //* Component Lifecycle Methods *
    //*******************************
    class_1.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.load()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    //******************************
    //* Private Method Definitions *
    //******************************
    class_1.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.window = window;
                //    this._element = this.el.shadowRoot;
                this.parseSlotStyleProp(this.slotstyle ? this.slotstyle : null);
                // check if there is a ion-menu and dismiss it
                this._menu = document.querySelector('ion-menu');
                if (this._menu !== null) {
                    this._menu.setAttribute('disabled', 'true');
                }
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var divs, slotDivEl, erEl;
            return __generator(this, function (_a) {
                divs = Array.from(this.el.querySelectorAll('div'));
                if (divs && divs.length > 0) {
                    slotDivEl = divs.filter(function (item) {
                        return (item.slot && item.slot === 'toprint') || item.outerHTML.indexOf('slot="toprint"') !== -1 ? item : null;
                    });
                    this._error = slotDivEl && slotDivEl.length > 0 ? null : "Error: slot name toprint doesn't exist";
                    if (this._error === null) {
                        erEl = this.el.shadowRoot.querySelector('#error-div');
                        erEl.style.setProperty('display', 'none');
                        slotDivEl[0].innerHTML = this.innerSlotStyle !== null ? this.innerSlotStyle + slotDivEl[0].innerHTML : slotDivEl[0].innerHTML;
                        this.printPageReady.emit();
                    }
                }
                else {
                    this._error = "Error: slot name toprint doesn't exist";
                }
                return [2 /*return*/];
            });
        });
    };
    //*************************
    //* Rendering JSX Element *
    //*************************
    class_1.prototype.render = function () {
        /*
        let toRender: any[] = [];
        if (this._error !== null) {
          toRender = [...toRender,
            <div id="error-div">{this._error}</div>
          ];
        } else {
          toRender = [...toRender,
            <slot name='toprint'></slot>
          ];
        }
        */
        return (h(Host, null, h("div", { id: "error-div" }, "Error: slot name toprint doesn't exist"), h("slot", { name: 'toprint' })));
    };
    Object.defineProperty(class_1.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "watchers", {
        get: function () {
            return {
                "slotstyle": ["parseSlotStyleProp"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "style", {
        get: function () { return ":host{display:block}#error-div{width:100%;height:50px;background-color:#d6a0a0;display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:center;justify-content:center;font-size:1.3rem;color:#000}"; },
        enumerable: true,
        configurable: true
    });
    return class_1;
}());
export { JeepHtmlToprint as jeep_html_toprint };
