// @jeepq/core/dist/esm-es5/jeep-stretchy-header.entry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-f19ae6a4.js';
import { f as getValueFromCss, h as cssVar } from './common-61073bba.js';
var JeepStretchyHeader = /** @class */ (function () {
    function class_1(hostRef) {
        registerInstance(this, hostRef);
        //************************
        //* Property Definitions *
        //************************
        /**
         * The Header Height
         */
        this.headerheight = "150px";
        /**
         * The Header Background Linear Gtradient if any and Url
         *  headerbackground="linear-gradient(to bottom, rgba(0,0,0,0), rgba(0,0,0,0.75)) ,
         *  url(https://raw.githubusercontent.com/jepiqueau/jeep/master/assets/images/italy-mountains.jpeg)"
         *
         */
        this.headerbackground = "url(https://raw.githubusercontent.com/jepiqueau/jeep/master/assets/images/italy-mountains.jpeg)";
        /**
         * The blur effect
         */
        this.headerbackgroundblur = false;
        /**
         * The Navbar Contrast Color
         */
        this.toolbarcontrastcolor = "#ffffff";
        this.innerBlur = false;
        this._mouseStart = false;
        this._ptStart = {};
        this._isScroll = false;
        this._scrollTop = 0;
        this._prevScrollTop = -1;
        this.onStretchyHeaderToolbar = createEvent(this, "jeepStretchyHeaderToolbar", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    class_1.prototype.parseHeaderHeight = function (newValue) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, headHeight;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = this;
                        if (!newValue) return [3 /*break*/, 2];
                        return [4 /*yield*/, getValueFromCss(newValue, 'y')];
                    case 1:
                        _b = _c.sent();
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, getValueFromCss("150px", 'y')];
                    case 3:
                        _b = _c.sent();
                        _c.label = 4;
                    case 4:
                        _a._heightIni = _b;
                        headHeight = cssVar(this.el, "--jeep-stretchy-header-height", this._heightIni + "px");
                        this._height = this._heightIni;
                        this.innerHeadHeight = headHeight;
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.parseHeaderBackground = function (newValue) {
        var url = newValue ? newValue : "url(https://raw.githubusercontent.com/jepiqueau/jeep/master/assets/images/lake-sunset-twilight.jpeg)";
        var backUrl = cssVar(this.el, "--jeep-stretchy-header-image", "" + url);
        this.innerUrl = backUrl;
    };
    class_1.prototype.parseBlur = function (newValue) {
        this.innerBlur = newValue ? newValue : false;
    };
    class_1.prototype.parseToolbarContrastColor = function (newValue) {
        this.innerContrastCol = newValue ? newValue : "#ffffff";
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_1.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    //*******************************
    //* Component Lifecycle Methods *
    //*******************************
    class_1.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._window = window;
                        return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._header = this.el.shadowRoot.querySelector('.stretchy-header');
                        //    this._header.style.setProperty('height',this.innerHeadHeight);
                        this._headerBackground = this.el.shadowRoot.querySelector('.stretchy-header-background');
                        //    this._headerBackground.style.setProperty("background-image",this.innerUrl);
                        this._ionHeader = this.el.querySelector('ion-header');
                        this._ionToolbar = this._ionHeader.querySelector('ion-toolbar');
                        this._ionContent = this.el.parentNode.querySelector('ion-content');
                        if (!this._ionHeader) return [3 /*break*/, 2];
                        return [4 /*yield*/, this._setHeader()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!this._ionToolbar) return [3 /*break*/, 4];
                        return [4 /*yield*/, this._setToolbar()];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        if (!this._ionContent) return [3 /*break*/, 6];
                        return [4 /*yield*/, this._setContent()];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6:
                        if (this._header.classList.contains("back"))
                            this._header.classList.remove("back");
                        if (this.innerBlur && this._headerBackground.classList.contains("blur"))
                            this._headerBackground.classList.remove("blur");
                        return [2 /*return*/];
                }
            });
        });
    };
    //******************************
    //* Private Method Definitions *
    //******************************
    class_1.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._document = this._window.document;
                this._root = this._document.documentElement;
                this.parseHeaderHeight(this.headerheight ? this.headerheight : "20%");
                this.parseHeaderBackground(this.headerbackground ? this.headerbackground : "url(https://ununsplash.imgix.net/photo-1421091242698-34f6ad7fc088?fit=crop&fm=jpg&h=650&q=75&w=950)");
                this.parseBlur(this.headerbackgroundblur ? this.headerbackgroundblur : false);
                this.parseToolbarContrastColor(this.toolbarcontrastcolor ? this.toolbarcontrastcolor : null);
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._setHeader = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._ionHeader.style.setProperty("position", "absolute");
                this._ionHeader.style.setProperty("top", "0");
                this._ionHeader.style.setProperty("left", "0");
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._setToolbar = function () {
        return __awaiter(this, void 0, void 0, function () {
            var color;
            return __generator(this, function (_a) {
                color = this._ionToolbar.color;
                this._ionColor = cssVar(this._root, "--ion-color-" + color);
                this._ionColorContrast = cssVar(this._root, "--ion-color-" + color + "-contrast");
                this._minHeight = this._ionToolbar.classList.contains('ios') ? 44 : 56;
                this._setTranslucent('add');
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._setContent = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this._ionContent.scrollEvents = true;
                this._ionContent.addEventListener('ionScrollStart', function () { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                if (!(this._height < this._heightIni)) return [3 /*break*/, 2];
                                this._isScroll = true;
                                this._mouseStart = false;
                                if (!(this._prevScrollTop === -1)) return [3 /*break*/, 2];
                                this._prevScrollTop = 0;
                                return [4 /*yield*/, this._ionContent.scrollToTop()];
                            case 1:
                                _a.sent();
                                _a.label = 2;
                            case 2: return [2 /*return*/];
                        }
                    });
                }); });
                this._ionContent.addEventListener('ionScroll', function (ev) { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                if (!(this._height < this._heightIni)) return [3 /*break*/, 4];
                                if (!this._isScroll)
                                    this._isScroll = true;
                                this._scrollTop = ev.detail.scrollTop;
                                this._height -= this._scrollTop;
                                this._height = this._height <= this._minHeight ? this._minHeight : this._height;
                                if (!(this._height > this._minHeight)) return [3 /*break*/, 2];
                                this._setTranslucent('add');
                                return [4 /*yield*/, this._ionContent.scrollToTop()];
                            case 1:
                                _a.sent();
                                this._prevScrollTop = -1;
                                return [3 /*break*/, 3];
                            case 2:
                                this._setTranslucent('remove');
                                _a.label = 3;
                            case 3:
                                this._header.style.setProperty('height', this._height + "px");
                                if (this._prevScrollTop !== -1 && this._scrollTop === 0) {
                                    this._setTranslucent('remove');
                                    this._isScroll = false;
                                    this._mouseStart = true;
                                }
                                this._prevScrollTop = this._scrollTop;
                                _a.label = 4;
                            case 4: return [2 /*return*/];
                        }
                    });
                }); });
                this._ionContent.addEventListener('ionScrollEnd', function () {
                    _this._isScroll = false;
                });
                this._ionContent.addEventListener('touchstart', function (ev) {
                    if (!_this._isScroll && _this._prevScrollTop <= 0) {
                        _this._handleStart(ev);
                    }
                });
                this._ionContent.addEventListener('touchmove', function (ev) {
                    _this._handleMove(ev);
                });
                this._ionContent.addEventListener('touchend', function () {
                    if (_this._mouseStart) {
                        _this._handleEnd();
                    }
                });
                this._ionContent.addEventListener('mousedown', function (ev) {
                    if (!_this._isScroll && _this._prevScrollTop <= 0) {
                        _this._handleStart(ev);
                    }
                });
                this._ionContent.addEventListener('mousemove', function (ev) {
                    _this._handleMove(ev);
                });
                this._ionContent.addEventListener('mouseup', function () {
                    if (_this._mouseStart) {
                        _this._handleEnd();
                    }
                });
                this._ionContent.addEventListener('mouseleave', function () {
                    if (_this._mouseStart) {
                        _this._handleEnd();
                    }
                });
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._setTranslucent = function (mode) {
        if (mode === 'add') {
            this.onStretchyHeaderToolbar.emit({
                color: this._ionColor + "00",
                contrastColor: this.innerContrastCol
            });
        }
        else if (mode === 'remove') {
            this.onStretchyHeaderToolbar.emit({
                color: this._ionColor,
                contrastColor: this._ionColorContrast
            });
        }
    };
    class_1.prototype._handleStart = function (ev) {
        this._ptStart = ev.touches && ev.touches[0] ? { x: ev.touches[0].pageX, y: ev.touches[0].pageY } : { x: ev.pageX, y: ev.pageY };
        if (!this._isScroll && this._height >= this._minHeight) {
            this._mouseStart = true;
        }
    };
    class_1.prototype._handleMove = function (ev) {
        return __awaiter(this, void 0, void 0, function () {
            var pt, dy;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        pt = ev.touches && ev.touches[0] ? { x: ev.touches[0].pageX, y: ev.touches[0].pageY } : { x: ev.pageX, y: ev.pageY };
                        dy = pt.y - this._ptStart.y;
                        this._ptStart = { x: pt.x, y: pt.y };
                        if (!(this._mouseStart && !this._isScroll)) return [3 /*break*/, 2];
                        this._height += dy;
                        if (dy > 0) {
                            this._setTranslucent('add');
                            if (this._height > this._heightIni + 2) {
                                if (this.innerBlur && this._headerBackground.classList.contains('unblur'))
                                    this._headerBackground.classList.remove("unblur");
                                if (this._header.classList.contains('back'))
                                    this._header.classList.remove('back');
                                if (this.innerBlur && !this._headerBackground.classList.contains('blur'))
                                    this._headerBackground.classList.add("blur");
                            }
                        }
                        else {
                            if (this.innerBlur && this._headerBackground.classList.contains('blur'))
                                this._headerBackground.classList.remove("blur");
                            if (this.innerBlur && !this._headerBackground.classList.contains('unblur'))
                                this._headerBackground.classList.add("unblur");
                            if (this._height <= this._heightIni) {
                                if (this._header.classList.contains('back'))
                                    this._header.classList.remove('back');
                            }
                            if (this._height <= this._minHeight) {
                                this._height = this._minHeight;
                                this._setTranslucent('remove');
                            }
                            else {
                                this._setTranslucent('add');
                            }
                        }
                        this._header.style.setProperty('height', this._height + "px");
                        if (!(this._height >= this._heightIni)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this._ionContent.scrollToTop()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._handleEnd = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this._isScroll) {
                    if (this._height >= this._heightIni) {
                        if (!this._header.classList.contains('back'))
                            this._header.classList.add('back');
                        if (this.innerBlur && this._headerBackground.classList.contains('blur'))
                            this._headerBackground.classList.remove("blur");
                        if (this.innerBlur && !this._headerBackground.classList.contains('unblur'))
                            this._headerBackground.classList.add("unblur");
                        this._height = this._heightIni;
                        this._header.style.setProperty('height', this._height + "px");
                    }
                    else {
                        if (this._header.classList.contains('back'))
                            this._header.classList.remove('back');
                        if (this._height <= this._minHeight + 2) {
                            this._height = this._minHeight;
                            this._setTranslucent('remove');
                            if (this.innerBlur && this._headerBackground.classList.contains("blur"))
                                this._headerBackground.classList.remove("blur");
                            if (this.innerBlur && this._headerBackground.classList.contains('unblur'))
                                this._headerBackground.classList.remove("unblur");
                        }
                        this._header.style.setProperty('height', this._height + "px");
                    }
                }
                if (this.innerBlur && this._headerBackground.classList.contains('unblur'))
                    this._headerBackground.classList.remove("unblur");
                this._mouseStart = false;
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype.render = function () {
        return (h(Host, null, h("div", { class: "stretchy-header" }, h("div", { class: "stretchy-header-background" }), h("slot", null))));
    };
    Object.defineProperty(class_1.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "watchers", {
        get: function () {
            return {
                "headerheight": ["parseHeaderHeight"],
                "headerbackground": ["parseHeaderBackground"],
                "headerbackgroundblur": ["parseBlur"],
                "toolbarcontrastcolor": ["parseToolbarContrastColor"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "style", {
        get: function () { return ":host{display:block;--jeep-stretchy-header-image:url(https://raw.githubusercontent.com/jepiqueau/jeep/master/assets/images/italy-mountains.jpeg);--jeep-stretchy-header-height:150px}.stretchy-header{height:var(--jeep-stretchy-header-height)}.stretchy-header-background{background-image:var(--jeep-stretchy-header-image);background-size:cover;background-position:50%;height:100%}.stretchy-header.back{-webkit-transition:height .4s linear;transition:height .4s linear}.stretchy-header-background.blur{transition:all 1s ease-in-out;-webkit-transition:all 1s ease-in-out;filter:blur(6px);-webkit-filter:blur(6px)}.stretchy-header-background.unblur{transition:all .4s ease;-webkit-transition:all .4s ease;filter:blur(0);-webkit-filter:blur(0)}"; },
        enumerable: true,
        configurable: true
    });
    return class_1;
}());
export { JeepStretchyHeader as jeep_stretchy_header };
