// @jeepq/core/dist/esm-es5/jeep-fullscreen_4.entry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-f19ae6a4.js';
var JeepFullscreen = /** @class */ (function () {
    function class_1(hostRef) {
        registerInstance(this, hostRef);
        //************************
        //* Property Definitions *
        //************************
        //*****************************
        //* Watch on Property Changes *
        //*****************************
        //************************
        //* State Definitions *
        //************************
        this.visibility = 'visible';
        this.fullscreen = false;
        this._fullscreenIcons = {
            request: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M396.795 396.8H320V448h128V320h-51.205zM396.8 115.205V192H448V64H320v51.205zM115.205 115.2H192V64H64v128h51.205zM115.2 396.795V320H64v128h128v-51.205z"/></svg>',
            exit: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M64 371.2h76.795V448H192V320H64v51.2zm76.795-230.4H64V192h128V64h-51.205v76.8zM320 448h51.2v-76.8H448V320H320v128zm51.2-307.2V64H320v128h128v-51.2h-76.8z"/></svg>'
        };
        this._fullscreenChange = false;
        this.onFullscreenRequest = createEvent(this, "jeepFullscreenRequest", 7);
        this.onFullscreenExit = createEvent(this, "jeepFullscreenExit", 7);
        this.onFullscreenChange = createEvent(this, "jeepFullscreenChange", 7);
    }
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    class_1.prototype.handleFullScreenChange = function () {
        if (this.fullscreen && this._fullscreenChange) {
            this.fullscreen = false;
            this.onFullscreenChange.emit();
        }
        this._fullscreenChange = !this._fullscreenChange;
    };
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_1.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    /**
     * Set the Fullscreen component.
     */
    class_1.prototype.setFullscreen = function () {
        return Promise.resolve(this._setFullscreen());
    };
    /**
     * Set the Fullscreen Visibility (visible/hidden)
     */
    class_1.prototype.setJeepFullscreenVisibility = function (state) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (state && state.visibility) {
                    if (state.visibility === 'hidden') {
                        if (!this._screenContainerEl.classList.contains('hidden'))
                            this._screenContainerEl.classList.add('hidden');
                    }
                    else {
                        if (this._screenContainerEl.classList.contains('hidden'))
                            this._screenContainerEl.classList.remove('hidden');
                    }
                    this.visibility = state.visibility;
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Request Fullscreen
     */
    class_1.prototype.fullscreenRequest = function (elem) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (elem.requestFullscreen) {
                    elem.requestFullscreen();
                }
                else if (elem.webkitRequestFullscreen) {
                    elem.webkitRequestFullscreen();
                }
                else if (elem.mozRequestFullScreen) {
                    elem.mozRequestFullScreen();
                }
                else if (elem.msRequestFullscreen) {
                    elem.msRequestFullscreen();
                }
                else {
                    this.fullscreen = false;
                    console.log('Fullscreen API is not supported.');
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Exit Fullscreen
     */
    class_1.prototype.fullscreenExit = function () {
        return __awaiter(this, void 0, void 0, function () {
            var doc;
            return __generator(this, function (_a) {
                doc = document;
                if (doc && doc.exitFullscreen) {
                    doc.exitFullscreen();
                }
                else if (doc && doc.webkitExitFullscreen) {
                    doc.webkitExitFullscreen();
                }
                else if (doc && doc.mozExitFullscreen) {
                    doc.mozExitFullscreen();
                }
                else if (doc && doc.msExitFullscreen) {
                    doc.msExitFullscreen();
                }
                else {
                    console.log('Fullscreen API is not supported.');
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Is Fullscreen
     */
    class_1.prototype.isFullscreen = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.fullscreen];
            });
        });
    };
    class_1.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.setFullscreen()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._element = this.el.shadowRoot;
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype._setFullscreen = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this._screenContainerEl = this._element.querySelector(".fullscreen-container");
                this._fullscreenEl = this._screenContainerEl.querySelector('.fullscreen-button');
                if (this._fullscreenEl) {
                    this._fullscreenEl.innerHTML = this._fullscreenIcons.request;
                    this._fullscreenEl.addEventListener('click', function () { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            if (this._fullscreenEl.classList.contains('fullscreen-button') && this._fullscreenEl.classList.contains('request')) {
                                this.fullscreen = true;
                                this.onFullscreenRequest.emit();
                            }
                            else {
                                if (this.fullscreen) {
                                    this.fullscreen = false;
                                    this.onFullscreenExit.emit();
                                }
                            }
                            return [2 /*return*/];
                        });
                    }); }, false);
                }
                return [2 /*return*/];
            });
        });
    };
    class_1.prototype.render = function () {
        if (this._fullscreenEl) {
            this._fullscreenEl.innerHTML = this.fullscreen ? this._fullscreenIcons.exit : this._fullscreenIcons.request;
        }
        return (h(Host, null, h("div", { class: "fullscreen-container" }, !this.fullscreen
            ? h("div", { class: "fullscreen-button request" })
            : h("div", { class: "fullscreen-button exit" }))));
    };
    Object.defineProperty(class_1.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_1, "style", {
        get: function () { return ":host{--fullscreen-top:10px;--fullscreen-left:10px;--fullscreen-width:100%;--fullscreen-button-width:50px;--fullscreen-button-height:50px;--fullscreen-button-icon-color:#222428}.fullscreen-container{position:absolute;top:var(--fullscreen-top);left:var(--fullscreen-left);width:var(--fullscreen-width);height:auto;z-index:12;visibility:visible}.fullscreen-button{display:inline-block;width:var(--fullscreen-button-width,20px);height:var(--fullscreen-button-height,20px);fill:var(--fullscreen-button-icon-color)}.hidden{visibility:hidden}"; },
        enumerable: true,
        configurable: true
    });
    return class_1;
}());
var icons = {
    "arrow-circle": {
        back: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M256 48C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48zm43.4 289.1c7.5 7.5 7.5 19.8 0 27.3-3.8 3.8-8.7 5.6-13.6 5.6s-9.9-1.9-13.7-5.7l-94-94.3c-6.9-7.6-6.7-19.3.6-26.6l95.4-95.7c7.5-7.5 19.7-7.6 27.3 0 7.5 7.5 7.6 19.7 0 27.3l-81.9 81 79.9 81.1z"/></svg>',
        forward: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M48 256c0 114.9 93.1 208 208 208s208-93.1 208-208S370.9 48 256 48 48 141.1 48 256zm244.5 0l-81.9-81.1c-7.5-7.5-7.5-19.8 0-27.3s19.8-7.5 27.3 0l95.4 95.7c7.3 7.3 7.5 19.1.6 26.6l-94 94.3c-3.8 3.8-8.7 5.7-13.7 5.7-4.9 0-9.9-1.9-13.6-5.6-7.5-7.5-7.6-19.7 0-27.3l79.9-81z"/></svg>'
    },
    "arrow-round": {
        back: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M401.4 224h-214l83-79.4c11.9-12.5 11.9-32.7 0-45.2s-31.2-12.5-43.2 0L89 233.4c-6 5.8-9 13.7-9 22.4v.4c0 8.7 3 16.6 9 22.4l138.1 134c12 12.5 31.3 12.5 43.2 0 11.9-12.5 11.9-32.7 0-45.2l-83-79.4h214c16.9 0 30.6-14.3 30.6-32 .1-18-13.6-32-30.5-32z"/></svg>',
        forward: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M284.9 412.6l138.1-134c6-5.8 9-13.7 9-22.4v-.4c0-8.7-3-16.6-9-22.4l-138.1-134c-12-12.5-31.3-12.5-43.2 0-11.9 12.5-11.9 32.7 0 45.2l83 79.4h-214c-17 0-30.7 14.3-30.7 32 0 18 13.7 32 30.6 32h214l-83 79.4c-11.9 12.5-11.9 32.7 0 45.2 12 12.5 31.3 12.5 43.3 0z"/></svg>'
    }
};
function getIcons(name) {
    if (name === 'arrow-circle' || name === 'arrow-round')
        return icons[name];
    return null;
}
var JeepNavigation = /** @class */ (function () {
    function class_2(hostRef) {
        registerInstance(this, hostRef);
        this.visibility = 'visible';
        this._prevDisabled = true;
        this._nextDisabled = false;
        this.onNavigationPrev = createEvent(this, "jeepNavigationPrev", 7);
        this.onNavigationNext = createEvent(this, "jeepNavigationNext", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    class_2.prototype.parseNameProp = function (newValue) {
        if (newValue)
            this.innerName = newValue;
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_2.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    /**
     * Set the navigation component.
     */
    class_2.prototype.setNavigation = function () {
        return Promise.resolve(this._setNavigation());
    };
    /**
     * Set the Navigation Visibility (visible/hidden)
     */
    class_2.prototype.setJeepNavigationVisibility = function (state) {
        if (state && state.visibility) {
            this.visibility = state.visibility;
            if (this.visibility === "visible") {
                this._navContainerEl.classList.remove('hidden');
            }
            else {
                this._navContainerEl.classList.add('hidden');
            }
        }
        return;
    };
    /**
     * Set the Navigation previous button to enable/disable
     */
    class_2.prototype.setJeepNavigationPrevDisabled = function (state) {
        this._prevDisabled = state.disabled;
        this._setPrevDisabled();
        return;
    };
    /**
     * Set the Navigation next button to enable/disable
     */
    class_2.prototype.setJeepNavigationNextDisabled = function (state) {
        this._nextDisabled = state.disabled;
        this._setNextDisabled();
        return;
    };
    /**
     * Get icon name
     */
    class_2.prototype.getJeepNavigationIconFamily = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this._icons !== null ? this.innerName : null];
            });
        });
    };
    class_2.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_2.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.setNavigation()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_2.prototype.componentDidUpdate = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.visibility === "visible")) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.setNavigation()];
                    case 1:
                        _a.sent();
                        this._setPrevDisabled();
                        this._setNextDisabled();
                        return [3 /*break*/, 3];
                    case 2:
                        this._navPrevEl = null;
                        this._navNextEl = null;
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    class_2.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._element = this.el.shadowRoot;
                this.parseNameProp(this.name ? this.name : "arrow-circle");
                return [2 /*return*/];
            });
        });
    };
    class_2.prototype._setNavigation = function () {
        return __awaiter(this, void 0, void 0, function () {
            var svgEls, i;
            var _this = this;
            return __generator(this, function (_a) {
                this._navContainerEl = this._element.querySelector(".navigation-container");
                this._navPrevEl = this._navContainerEl.querySelector('.navigation-prev');
                this._navNextEl = this._navContainerEl.querySelector('.navigation-next');
                this._navPrevEl.innerHTML = this._prevButton;
                this._navNextEl.innerHTML = this._nextButton;
                svgEls = Array.from(this._navContainerEl.querySelectorAll('svg'));
                for (i = 0; i < svgEls.length; i++) {
                    svgEls[i].classList.add('icon');
                }
                this._navPrevEl.addEventListener('click', function () {
                    if (!_this._navPrevEl.classList.contains('disabled'))
                        _this.onNavigationPrev.emit();
                }, false);
                this._navNextEl.addEventListener('click', function () {
                    if (!_this._navNextEl.classList.contains('disabled'))
                        _this.onNavigationNext.emit();
                }, false);
                return [2 /*return*/];
            });
        });
    };
    class_2.prototype._setPrevDisabled = function () {
        if (this._navPrevEl) {
            if (this._prevDisabled) {
                if (!this._navPrevEl.classList.contains('disabled'))
                    this._navPrevEl.classList.add('disabled');
            }
            else {
                if (this._navPrevEl.classList.contains('disabled'))
                    this._navPrevEl.classList.remove('disabled');
            }
        }
    };
    class_2.prototype._setNextDisabled = function () {
        if (this._navNextEl) {
            if (this._nextDisabled) {
                if (!this._navNextEl.classList.contains('disabled'))
                    this._navNextEl.classList.add('disabled');
            }
            else {
                if (this._navNextEl.classList.contains('disabled'))
                    this._navNextEl.classList.remove('disabled');
            }
        }
    };
    class_2.prototype.render = function () {
        this._icons = getIcons(this.innerName);
        this._prevButton = this._icons !== null ? this._icons.back : null;
        this._nextButton = this._icons !== null ? this._icons.forward : null;
        return (h(Host, null, h("div", { class: "navigation-container" }, this.visibility === "visible"
            ?
                h("div", { class: "navigation-wrapper" }, h("button", { class: 'navigation-prev icon disabled' }), h("button", { class: 'navigation-next icon disabled' }))
            : null)));
    };
    Object.defineProperty(class_2.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_2, "watchers", {
        get: function () {
            return {
                "name": ["parseNameProp"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_2, "style", {
        get: function () { return ":host{--navigation-top:50%;--navigation-left:0;--navigation-width:100vw;--navigation-button-width:50px;--navigation-button-height:50px;--navigation-padding-horizontal:10px;--navigation-button-icon-color:#222428}.navigation-container{position:absolute;top:var(--navigation-top);left:var(--navigation-left);margin-top:calc(-1 * var(--navigation-button-height) / 2);z-index:10;width:var(--navigation-width);height:auto;display:-ms-flexbox;display:flex;padding:0 var(--navigation-padding-horizontal);visibility:visible}.navigation-wrapper{width:calc(100% - 2 * var(--navigation-padding-horizontal));height:auto}.icon{position:relative;width:var(--navigation-button-width);height:var(--navigation-button-height);fill:var(--navigation-button-icon-color);visibility:visible}.navigation-prev{left:0}.navigation-next,.navigation-prev{padding:0;border-width:0;background:transparent;border-color:transparent;border-radius:100%;opacity:1}.navigation-next{left:calc(100% - 2 * var(--navigation-button-width))}.disabled{opacity:.2}.hidden{visibility:hidden}"; },
        enumerable: true,
        configurable: true
    });
    return class_2;
}());
var JeepPagination = /** @class */ (function () {
    function class_3(hostRef) {
        registerInstance(this, hostRef);
        this.visibility = 'visible';
        this.nbItems = 0;
        this.activeIndex = 0;
        this.onPaginationIndex = createEvent(this, "jeepPaginationIndex", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    class_3.prototype.parseNDisplayProp = function (newValue) {
        if (newValue) {
            var val = newValue;
            this.innerNDisplay = val % 2 ? val : val + 1;
            this._delta = Math.floor(this.innerNDisplay / 2);
        }
    };
    class_3.prototype.parseDirectionProp = function (newValue) {
        var dirs = ["horizontal", "vertical"];
        this.innerDirection = dirs.includes(newValue) ? newValue : "horizontal";
    };
    class_3.prototype.parseClickableProp = function (newValue) {
        this.innerClickable = newValue ? newValue : false;
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_3.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    /**
     * Set the Pagination Visibility (visible/hidden)
     */
    class_3.prototype.setJeepPaginationVisibility = function (state) {
        if (state && state.visibility) {
            this.visibility = state.visibility;
        }
        return;
    };
    /**
     * Set the Pagination Slides Number
     */
    class_3.prototype.setJeepPaginationSlidesNumber = function (state) {
        if (state && state.slides) {
            this.nbItems = state.slides;
        }
        return;
    };
    /**
     * Set the Pagination Active Index
     */
    class_3.prototype.setJeepPaginationActiveIndex = function (state) {
        if (state && state.activeIndex) {
            this.activeIndex = Number(state.activeIndex);
        }
        return;
    };
    /**
     * Get the Pagination Number of bullets displayed
     */
    class_3.prototype.getJeepPaginationBulletNumber = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.innerNDisplay];
            });
        });
    };
    class_3.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_3.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._element = this.el.shadowRoot;
                this.parseNDisplayProp(this.ndisplay ? this.ndisplay : 5);
                this.parseDirectionProp(this.direction ? this.direction : "horizontal");
                this.parseClickableProp(this.clickable ? this.clickable : false);
                return [2 /*return*/];
            });
        });
    };
    class_3.prototype._handleClick = function (i) {
        if (this.innerClickable) {
            this.onPaginationIndex.emit({ index: i.toString() });
        }
    };
    class_3.prototype._setPaginationBullets = function (nbItems) {
        var _this = this;
        var bullets = [];
        var initIndex = this.activeIndex - this._delta;
        var classname = "pagination-bullet pagination-bullet-" + this.innerDirection;
        var _loop_1 = function (i) {
            var appliedClass = classname;
            if (initIndex + i < 0) {
                appliedClass = classname + " hidden-bullet";
            }
            else if (initIndex + i === this_1.activeIndex) {
                appliedClass = classname + " active-bullet";
            }
            else if (initIndex + i > nbItems - 1) {
                appliedClass = classname + " hidden-bullet";
            }
            bullets = __spreadArrays(bullets, [h("span", { class: appliedClass, onClick: function () { return _this._handleClick(initIndex + i); } })]);
        };
        var this_1 = this;
        for (var i = 0; i < this.innerNDisplay; i++) {
            _loop_1(i);
        }
        //        this.initialIndex = initIndex;
        return bullets;
    };
    class_3.prototype.render = function () {
        var paginationBullets = this.nbItems > 0 ? this._setPaginationBullets(this.nbItems) : [];
        var containerClass = "pagination-container pagination-container-" + this.innerDirection;
        return (h(Host, null, h("div", { class: containerClass }, this.visibility === "visible"
            ?
                h("div", { class: "pagination-wrapper" }, paginationBullets)
            : null)));
    };
    Object.defineProperty(class_3.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_3, "watchers", {
        get: function () {
            return {
                "ndisplay": ["parseNDisplayProp"],
                "direction": ["parseDirectionProp"],
                "clickable": ["parseClickableProp"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_3, "style", {
        get: function () { return ":host{--pagination-top:5%;--pagination-left:0;--pagination-width:100vw;--pagination-height:100vh;--bullet-background:#000;--bullet-opacity:0.5;--bullet-active-background:#ff8000;--bullet-active-opacity:1;--bullet-active-diameter:22px;--bullet-diameter:15px}.pagination-container{position:absolute;z-index:8;text-align:center;padding:0}.pagination-container-horizontal{top:var(--pagination-top);left:var(--pagination-left);margin-top:calc(-1 * var(--bullet-active-diameter) / 2);width:var(--pagination-width);height:auto}.pagination-container-vertical{top:var(--pagination-top);left:var(--pagination-left);width:calc(var(--bullet-active-diameter));height:var(--pagination-height);display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-align:center;align-items:center}.pagination-wrapper{margin:0 auto;width:100%;height:auto}.pagination-bullet{display:inline-block;width:var(--bullet-diameter);height:var(--bullet-diameter);border-radius:100%;opacity:var(--bullet-opacity);background:var(--bullet-background);visibility:visible}.pagination-bullet-horizontal{margin:0 4px}.pagination-bullet-vertical{margin:4px 0}.active-bullet{width:var(--bullet-active-diameter);height:var(--bullet-active-diameter);opacity:var(--bullet-active-opacity);background:var(--bullet-active-background)}.hidden,.hidden-bullet{visibility:hidden}"; },
        enumerable: true,
        configurable: true
    });
    return class_3;
}());
var JeepPlayControls = /** @class */ (function () {
    function class_4(hostRef) {
        registerInstance(this, hostRef);
        this.visibility = 'visible';
        this.play = false;
        this._skipForwardDisabled = true;
        this._skipBackwardDisabled = false;
        this.onPlayControlsIsBeginning = createEvent(this, "jeepPlayControlsIsBeginning", 7);
        this.onPlayControlsIsEnd = createEvent(this, "jeepPlayControlsIsEnd", 7);
        this.onPlayControlsCurrentIndex = createEvent(this, "jeepPlayControlsCurrentIndex", 7);
        this.onPlayControlsAutoplayStart = createEvent(this, "jeepPlayControlsAutoplayStart", 7);
        this.onPlayControlsAutoplayStop = createEvent(this, "jeepPlayControlsAutoplayStop", 7);
        this.onPlayControlsAutoplayPause = createEvent(this, "jeepPlayControlsAutoplayPause", 7);
        this.onPlayControlsAutoplaySkipBackward = createEvent(this, "jeepPlayControlsAutoplaySkipBackward", 7);
        this.onPlayControlsAutoplaySkipForward = createEvent(this, "jeepPlayControlsAutoplaySkipForward", 7);
    }
    //*****************************
    //* Watch on Property Changes *
    //*****************************
    class_4.prototype.parseDurationProp = function (newValue) {
        if (newValue)
            this.innerDuration = newValue;
    };
    class_4.prototype.parseNSlidesProp = function (newValue) {
        if (newValue !== 0) {
            this.innerNSlides = newValue;
            this.innerToSlide = this.innerNSlides - 1;
        }
        else {
            this.innerNSlides = 0;
            this.innerToSlide = 0;
        }
    };
    class_4.prototype.parseFromSlideProp = function (newValue) {
        if (newValue !== 0) {
            this.innerFromSlide = newValue;
        }
        else {
            this.innerFromSlide = 0;
        }
    };
    class_4.prototype.parseToSlideProp = function (newValue) {
        if (newValue !== 0) {
            this.innerToSlide = newValue;
        }
        else {
            this.innerToSlide = 0;
        }
    };
    //*******************************
    //* Listen to Event Definitions *
    //*******************************
    //**********************
    //* Method Definitions *
    //**********************
    /**
     * Init data from properties.
     */
    class_4.prototype.init = function () {
        return Promise.resolve(this._init());
    };
    /**
     * Set the PlayControls component.
     */
    class_4.prototype.setPlayControls = function () {
        return Promise.resolve(this._setPlayControls());
    };
    /**
     * Set the PlayControls Visibility (visible/hidden)
     */
    class_4.prototype.setJeepPlayControlsVisibility = function (state) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (state && state.visibility) {
                    if (state.visibility === 'hidden') {
                        if (!this._playContainerEl.classList.contains('hidden'))
                            this._playContainerEl.classList.add('hidden');
                    }
                    else {
                        if (this._playContainerEl.classList.contains('hidden'))
                            this._playContainerEl.classList.remove('hidden');
                    }
                    this.visibility = state.visibility;
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Set the PlayControls skipbackward icon to enable/disable
     */
    class_4.prototype.setJeepPlayControlsSkipBackwardDisabled = function (state) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._skipBackwardDisabled = state.disabled;
                this._setSkipBackwardDisabled();
                return [2 /*return*/];
            });
        });
    };
    /**
     * Set the PlayControls skipforward icon to enable/disable
     */
    class_4.prototype.setJeepPlayControlsSkipForwardDisabled = function (state) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._skipForwardDisabled = state.disabled;
                this._setSkipForwardDisabled();
                return [2 /*return*/];
            });
        });
    };
    /**
     * Get the PlayControls duration
     */
    class_4.prototype.getJeepPlayControlsDuration = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.innerDuration];
            });
        });
    };
    /**
     * Start autoplay
     */
    class_4.prototype.autoplayStart = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.play = true;
                this.onPlayControlsAutoplayStart.emit();
                this._playContent();
                return [2 /*return*/];
            });
        });
    };
    /**
     * Stop autoplay
     */
    class_4.prototype.autoplayStop = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.play = false;
                        this.onPlayControlsAutoplayStop.emit();
                        clearInterval(this._initial);
                        return [4 /*yield*/, this.setCurrentIndex(this.innerFromSlide)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Pause autoplay
     */
    class_4.prototype.autoplayPause = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.play = false;
                this.onPlayControlsAutoplayPause.emit();
                clearInterval(this._initial);
                return [2 /*return*/];
            });
        });
    };
    /**
     * Is Playing
     */
    class_4.prototype.isPlaying = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.play];
            });
        });
    };
    /**
     * Get PlayControls Current Index
     */
    class_4.prototype.getCurrentIndex = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this._currentIndex];
            });
        });
    };
    /**
     * Set PlayControls Active Index
     */
    class_4.prototype.setActiveIndexAndPlay = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                clearInterval(this._initial);
                this.setCurrentIndex(index);
                this.autoplayStart();
                return [2 /*return*/];
            });
        });
    };
    /**
     * Set PlayControls Current Index
     */
    class_4.prototype.setCurrentIndex = function (index, notemit) {
        return __awaiter(this, void 0, void 0, function () {
            var willEmit;
            return __generator(this, function (_a) {
                willEmit = notemit ? !notemit : true;
                this._currentIndex = index;
                this.onPlayControlsCurrentIndex.emit({ index: this._currentIndex.toString() });
                this._setSkipButtons();
                if (willEmit && this._currentIndex === this.innerFromSlide)
                    this.onPlayControlsIsBeginning.emit();
                if (willEmit && this._currentIndex === this.innerToSlide)
                    this.onPlayControlsIsEnd.emit();
                return [2 /*return*/];
            });
        });
    };
    class_4.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_4.prototype.componentDidLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.setPlayControls()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_4.prototype._init = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._element = this.el.shadowRoot;
                        this.parseDurationProp(this.duration ? this.duration : 1000);
                        this.parseNSlidesProp(this.nslides ? this.nslides : 0);
                        this.parseFromSlideProp(this.fromslide ? this.fromslide : 0);
                        this.parseToSlideProp(this.toslide ? this.toslide : this.innerNSlides > 0 ? this.innerNSlides - 1 : 0);
                        return [4 /*yield*/, this.setCurrentIndex(this.innerFromSlide, true)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    class_4.prototype._setPlayControls = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this._playContainerEl = this._element.querySelector(".playcontrols-container");
                this._playSkipBackwardEl = this._playContainerEl.querySelector('.playcontrols-button-skip-backward');
                this._playSkipForwardEl = this._playContainerEl.querySelector('.playcontrols-button-skip-forward');
                this.setJeepPlayControlsSkipBackwardDisabled({ disabled: true });
                this.onPlayControlsIsBeginning.emit();
                if (this.innerToSlide > 0) {
                    this.setJeepPlayControlsSkipForwardDisabled({ disabled: false });
                }
                else {
                    this.setJeepPlayControlsSkipForwardDisabled({ disabled: true });
                    this.onPlayControlsIsEnd.emit();
                }
                return [2 /*return*/];
            });
        });
    };
    class_4.prototype._setSkipBackwardDisabled = function () {
        if (this._skipBackwardDisabled) {
            if (!this._playSkipBackwardEl.classList.contains('disabled'))
                this._playSkipBackwardEl.classList.add('disabled');
        }
        else {
            if (this._playSkipBackwardEl.classList.contains('disabled'))
                this._playSkipBackwardEl.classList.remove('disabled');
        }
    };
    class_4.prototype._setSkipForwardDisabled = function () {
        if (this._skipForwardDisabled) {
            if (!this._playSkipForwardEl.classList.contains('disabled'))
                this._playSkipForwardEl.classList.add('disabled');
        }
        else {
            if (this._playSkipForwardEl.classList.contains('disabled'))
                this._playSkipForwardEl.classList.remove('disabled');
        }
    };
    class_4.prototype._handleClick = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(event.target.classList.contains('playcontrols-button-play') && event.target.classList.contains('play'))) return [3 /*break*/, 3];
                        if (!(this.innerToSlide > 0 && this._currentIndex < this.innerToSlide)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.autoplayStart()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [3 /*break*/, 11];
                    case 3:
                        if (!event.target.classList.contains('playcontrols-button-skip-backward')) return [3 /*break*/, 6];
                        if (!(!this._playSkipBackwardEl.classList.contains('disabled') || this._currentIndex > this.innerFromSlide)) return [3 /*break*/, 5];
                        this.onPlayControlsAutoplaySkipBackward.emit();
                        return [4 /*yield*/, this.setActiveIndexAndPlay(this._currentIndex - 1)];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5: return [3 /*break*/, 11];
                    case 6:
                        if (!event.target.classList.contains('playcontrols-button-skip-forward')) return [3 /*break*/, 9];
                        if (!(!this._playSkipForwardEl.classList.contains('disabled') || this._currentIndex < this.innerToSlide)) return [3 /*break*/, 8];
                        this.onPlayControlsAutoplaySkipForward.emit();
                        return [4 /*yield*/, this.setActiveIndexAndPlay(this._currentIndex + 1)];
                    case 7:
                        _a.sent();
                        _a.label = 8;
                    case 8: return [3 /*break*/, 11];
                    case 9:
                        if (!this.play) return [3 /*break*/, 11];
                        return [4 /*yield*/, this.autoplayPause()];
                    case 10:
                        _a.sent();
                        _a.label = 11;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    class_4.prototype._playContent = function () {
        var _this = this;
        var index = 0;
        if (this.play) {
            this._initial = setInterval(function () {
                if (index === 0) {
                    index = _this._currentIndex + 1;
                }
                if (index > _this.innerToSlide) {
                    _this.play = false;
                    _this.setCurrentIndex(_this.innerFromSlide);
                    clearInterval(_this._initial);
                }
                else {
                    _this.setCurrentIndex(index);
                    index++;
                }
            }, this.innerDuration);
        }
        else {
            clearInterval(this._initial);
            return;
        }
    };
    class_4.prototype._setSkipButtons = function () {
        if (this._playSkipBackwardEl && this._playSkipForwardEl) {
            if (this._currentIndex === this.innerFromSlide) {
                this.setJeepPlayControlsSkipBackwardDisabled({ disabled: true });
            }
            else {
                this.setJeepPlayControlsSkipBackwardDisabled({ disabled: false });
            }
            if (this._currentIndex === this.innerToSlide) {
                this.setJeepPlayControlsSkipForwardDisabled({ disabled: true });
            }
            else {
                this.setJeepPlayControlsSkipForwardDisabled({ disabled: false });
            }
        }
    };
    class_4.prototype.render = function () {
        var _this = this;
        this._setSkipButtons();
        return (h(Host, null, h("div", { class: "playcontrols-container" }, h("div", { class: "playcontrols-button playcontrols-button-skip-backward disabled", onClick: function (event) { return _this._handleClick(event); } }), this.play
            ? h("div", { class: "playcontrols-button playcontrols-button-play pause", onClick: function (event) { return _this._handleClick(event); } })
            : h("div", { class: "playcontrols-button playcontrols-button-play play", onClick: function (event) { return _this._handleClick(event); } }), h("div", { class: "playcontrols-button playcontrols-button-skip-forward disabled", onClick: function (event) { return _this._handleClick(event); } }))));
    };
    Object.defineProperty(class_4.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_4, "watchers", {
        get: function () {
            return {
                "duration": ["parseDurationProp"],
                "nslides": ["parseNSlidesProp"],
                "fromslide": ["parseFromSlideProp"],
                "toslide": ["parseToSlideProp"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(class_4, "style", {
        get: function () { return ":host{--playcontrols-top:calc(100vh -20px);--playcontrols-left:0px;--playcontrols-width:100%;--playcontrols-button-width:50px;--playcontrols-button-height:50px}.playcontrols-button-play.pause{background-image:url(\"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20512%20512%22%3E%3Cpath%20d=%22M199.9%20416h-63.8c-4.5%200-8.1-3.6-8.1-8V104c0-4.4%203.6-8%208.1-8h63.8c4.5%200%208.1%203.6%208.1%208v304c0%204.4-3.6%208-8.1%208zM375.9%20416h-63.8c-4.5%200-8.1-3.6-8.1-8V104c0-4.4%203.6-8%208.1-8h63.8c4.5%200%208.1%203.6%208.1%208v304c0%204.4-3.6%208-8.1%208z%22/%3E%3C/svg%3E\")}.playcontrols-button-play.pause,.playcontrols-button-play.play{width:calc(var(--playcontrols-button-width,20px) * 1.4)!important;height:calc(var(--playcontrols-button-height,20px) * 1.4)!important;padding-bottom:0!important}.playcontrols-button-play.play{background-image:url(\"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20512%20512%22%3E%3Cpath%20d=%22M128%20104.3v303.4c0%206.4%206.5%2010.4%2011.7%207.2l240.5-151.7c5.1-3.2%205.1-11.1%200-14.3L139.7%2097.2c-5.2-3.3-11.7.7-11.7%207.1z%22/%3E%3C/svg%3E\")}.playcontrols-button-skip-forward{background-image:url(\"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20512%20512%22%3E%3Cpath%20d=%22M337%20100v137.8L108.1%2098.1C102.7%2095%2096%2098.8%2096%20105v302c0%206.2%206.7%2010%2012.1%206.9L337%20274.2V412c0%202.2%201.8%204%204%204h71c2.2%200%204-1.8%204-4V100c0-2.2-1.8-4-4-4h-71c-2.2%200-4%201.8-4%204z%22/%3E%3C/svg%3E\")}.playcontrols-button-skip-backward{background-image:url(\"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20512%20512%22%3E%3Cpath%20d=%22M175%20100v137.8L403.9%2098.1c5.3-3.1%2012.1.7%2012.1%206.9v302c0%206.2-6.7%2010-12.1%206.9L175%20274.2V412c0%202.2-1.8%204-4%204h-71c-2.2%200-4-1.8-4-4V100c0-2.2%201.8-4%204-4h71c2.2%200%204%201.8%204%204z%22/%3E%3C/svg%3E\")}.playcontrols-container{position:absolute;top:var(--playcontrols-top);left:var(--playcontrols-left);width:var(--playcontrols-width);height:auto;text-align:center;z-index:11;visibility:visible}.playcontrols-button{display:inline-block;width:var(--playcontrols-button-width,20px);height:var(--playcontrols-button-height,20px);padding-bottom:calc(var(--playcontrols-button-height,20px) * .2)}.disabled{opacity:.2}.hidden{visibility:hidden}"; },
        enumerable: true,
        configurable: true
    });
    return class_4;
}());
export { JeepFullscreen as jeep_fullscreen, JeepNavigation as jeep_navigation, JeepPagination as jeep_pagination, JeepPlayControls as jeep_play_controls };
